// rrule-temporal/dist/totext.cjs
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/totext.ts
var totext_exports = {};
__export(totext_exports, {
  toText: () => toText
});
module.exports = __toCommonJS(totext_exports);

// src/index.ts
var import_polyfill = require("@js-temporal/polyfill");
var allowedFreq = ["YEARLY", "MONTHLY", "WEEKLY", "DAILY", "HOURLY", "MINUTELY", "SECONDLY"];
var allowedWeekdays = ["MO", "TU", "WE", "TH", "FR", "SA", "SU"];
var weekdayToIsoDay = {
  MO: 1,
  TU: 2,
  WE: 3,
  TH: 4,
  FR: 5,
  SA: 6,
  SU: 7
};
var allowedFreqSet = new Set(allowedFreq);
var allowedWeekdaysSet = new Set(allowedWeekdays);
var byDayTokenRegex = new RegExp(`^([+-]?\\d{1,2})?(${allowedWeekdays.join("|")})$`);
var byDayWeekdaySuffixRegex = new RegExp(`(${allowedWeekdays.join("|")})$`);
var MS_PER_SECOND = 1e3;
var MS_PER_MINUTE = 60 * MS_PER_SECOND;
var MS_PER_HOUR = 60 * MS_PER_MINUTE;
var MS_PER_DAY = 24 * MS_PER_HOUR;
var MS_PER_WEEK = 7 * MS_PER_DAY;
var GREGORIAN_MONTH_LENGTHS = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
var GREGORIAN_WEEKDAY_OFFSETS = [0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4];
var NS_PER_MILLISECOND = BigInt(1e6);
var NS_PER_SECOND = BigInt(1e9);
var NS_PER_MINUTE = BigInt(60) * NS_PER_SECOND;
var NS_PER_HOUR = BigInt(60) * NS_PER_MINUTE;
var NS_PER_DAY = BigInt(24) * NS_PER_HOUR;
var NS_PER_WEEK = BigInt(7) * NS_PER_DAY;
function addIsoDays(dayOfWeek, deltaDays) {
  return (dayOfWeek - 1 + deltaDays % 7 + 7) % 7 + 1;
}
function extractWeekdayToken(token) {
  const m = token.toUpperCase().match(byDayWeekdaySuffixRegex);
  const weekday = m == null ? void 0 : m[1];
  if (!weekday || !allowedWeekdaysSet.has(weekday)) return null;
  return weekday;
}
function parseByDayToken(token) {
  const m = token.toUpperCase().match(byDayTokenRegex);
  if (!m) return null;
  const ord = m[1] ? parseInt(m[1], 10) : 0;
  const weekday = m[2];
  if (!weekday || !allowedWeekdaysSet.has(weekday)) return null;
  return { ord, weekday };
}
function isIcsOpts(opts) {
  return typeof opts.rruleString === "string";
}
function unfoldLine(foldedLine) {
  return foldedLine.replace(/\r?\n[ \t]/g, "");
}
function parseIcsDateTime(dateStr, tzid, valueType) {
  const isDate = valueType === "DATE" || !dateStr.includes("T");
  const isoDate = `${dateStr.slice(0, 4)}-${dateStr.slice(4, 6)}-${dateStr.slice(6, 8)}`;
  if (isDate) {
    return import_polyfill.Temporal.PlainDate.from(isoDate).toZonedDateTime({ timeZone: tzid });
  }
  if (dateStr.endsWith("Z")) {
    const iso = `${isoDate}T${dateStr.slice(9, 15)}Z`;
    return import_polyfill.Temporal.Instant.from(iso).toZonedDateTimeISO(tzid || "UTC");
  } else {
    const iso = `${isoDate}T${dateStr.slice(9)}`;
    return import_polyfill.Temporal.PlainDateTime.from(iso).toZonedDateTime(tzid);
  }
}
function parseDateLines(lines, linePrefix, defaultTzid) {
  const dates = [];
  const regex = new RegExp(`^${linePrefix}(?:;VALUE=([^;]+))?(?:;TZID=([^:]+))?:(.+)`, "i");
  for (const line of lines) {
    const match = line.match(regex);
    if (match) {
      const [, valueType, tzid, dateValuesStr] = match;
      const timezone = tzid || defaultTzid;
      const dateValues = dateValuesStr.split(",");
      dates.push(...dateValues.map((dateValue) => parseIcsDateTime(dateValue, timezone, valueType)));
    }
  }
  return dates;
}
function parseNumberArray(val, sort = false) {
  const arr = val.split(",").map((n) => parseInt(n, 10));
  if (sort) {
    return arr.sort((a, b) => a - b);
  }
  return arr;
}
function parseByMonthArray(val) {
  return val.split(",").map((tok) => {
    const t = tok.trim();
    if (/^\d+L$/i.test(t)) return t.toUpperCase();
    const n = parseInt(t, 10);
    return Number.isFinite(n) ? n : t;
  });
}
function parseRRuleString(input, targetTimezone, dtstart, strict = false) {
  var _a, _b, _c, _d, _e;
  const unfoldedInput = unfoldLine(input).trim();
  let parsedDtstart;
  let tzid = targetTimezone;
  let dtstartValueType = "DATE-TIME";
  let dtstartHasTzid = false;
  let dtstartIsUtc = false;
  let rruleLine;
  let exDate = [];
  let rDate = [];
  if (/^DTSTART/im.test(unfoldedInput)) {
    const lines = unfoldedInput.split(/\s+/);
    const dtLine = lines.find((line) => line.match(/^DTSTART/i));
    const rrLine = lines.find((line) => line.match(/^RRULE:/i));
    const exLines = lines.filter((line) => line.match(/^EXDATE/i));
    const rLines = lines.filter((line) => line.match(/^RDATE/i));
    const dtMatch = dtLine.match(/DTSTART(?:;VALUE=([^;]+))?(?:;TZID=([^:]+))?:(.+)/i);
    if (!dtMatch) throw new Error("Invalid DTSTART in ICS snippet");
    const [, valueType, dtTzid, dtValue] = dtMatch;
    const normalizedValueType = (valueType || ((dtValue == null ? void 0 : dtValue.includes("T")) ? "DATE-TIME" : "DATE")).toUpperCase();
    dtstartValueType = normalizedValueType === "DATE" ? "DATE" : "DATE-TIME";
    dtstartHasTzid = Boolean(dtTzid);
    dtstartIsUtc = Boolean(dtValue == null ? void 0 : dtValue.endsWith("Z"));
    const effectiveTzid = (_b = (_a = dtTzid != null ? dtTzid : targetTimezone) != null ? _a : tzid) != null ? _b : "UTC";
    parsedDtstart = parseIcsDateTime(dtValue, effectiveTzid, dtstartValueType);
    tzid = (_e = (_d = (_c = dtTzid != null ? dtTzid : parsedDtstart.timeZoneId) != null ? _c : targetTimezone) != null ? _d : tzid) != null ? _e : "UTC";
    rruleLine = rrLine;
    exDate = parseDateLines(exLines, "EXDATE", tzid != null ? tzid : "UTC");
    rDate = parseDateLines(rLines, "RDATE", tzid != null ? tzid : "UTC");
  } else {
    parsedDtstart = dtstart;
    rruleLine = unfoldedInput;
    if (parsedDtstart) {
      tzid = parsedDtstart.timeZoneId;
      dtstartValueType = "DATE-TIME";
      dtstartHasTzid = true;
      dtstartIsUtc = parsedDtstart.timeZoneId === "UTC";
    }
  }
  const parts = rruleLine ? rruleLine.replace(/^RRULE:/i, "").split(";") : [];
  const opts = {
    dtstart: parsedDtstart,
    tzid,
    exDate: exDate.length > 0 ? exDate : void 0,
    rDate: rDate.length > 0 ? rDate : void 0
  };
  let pendingSkip;
  for (const part of parts) {
    const [key, val] = part.split("=");
    if (!key) continue;
    switch (key.toUpperCase()) {
      case "RSCALE":
        if (val) {
          opts.rscale = val.toUpperCase();
          if (pendingSkip && !opts.skip) {
            opts.skip = pendingSkip;
            pendingSkip = void 0;
          }
        }
        break;
      case "SKIP": {
        const v = (val || "").toUpperCase();
        if (!["OMIT", "BACKWARD", "FORWARD"].includes(v)) {
          throw new Error(`Invalid SKIP value: ${val}`);
        }
        if (opts.rscale) {
          opts.skip = v;
        } else {
          pendingSkip = v;
        }
        break;
      }
      case "FREQ":
        opts.freq = val.toUpperCase();
        break;
      case "INTERVAL":
        opts.interval = parseInt(val, 10);
        break;
      case "COUNT":
        opts.count = parseInt(val, 10);
        break;
      case "UNTIL": {
        const untilHasTime = val.includes("T");
        if (dtstartValueType === "DATE") {
          if (untilHasTime) {
            throw new Error("UNTIL rule part MUST have the same value type as DTSTART");
          }
          opts.until = parseIcsDateTime(val, tzid || "UTC", "DATE");
          break;
        }
        if (!untilHasTime) {
          if (strict) {
            throw new Error("UNTIL rule part MUST have the same value type as DTSTART");
          }
          const localEndOfDay = parseIcsDateTime(val, tzid || "UTC", "DATE").with({
            hour: 23,
            minute: 59,
            second: 59,
            millisecond: 0,
            microsecond: 0,
            nanosecond: 0
          });
          const requiresUtc2 = dtstartHasTzid || dtstartIsUtc;
          opts.until = requiresUtc2 ? localEndOfDay.withTimeZone("UTC") : localEndOfDay;
          break;
        }
        const requiresUtc = dtstartHasTzid || dtstartIsUtc;
        if (requiresUtc && !val.endsWith("Z")) {
          throw new Error("UNTIL rule part MUST always be specified as a date with UTC time");
        }
        opts.until = parseIcsDateTime(val, tzid || "UTC", "DATE-TIME");
        break;
      }
      case "BYHOUR":
        opts.byHour = parseNumberArray(val, true);
        break;
      case "BYMINUTE":
        opts.byMinute = parseNumberArray(val, true);
        break;
      case "BYSECOND":
        opts.bySecond = parseNumberArray(val, true);
        break;
      case "BYDAY":
        opts.byDay = val.split(",").map((token) => token.toUpperCase());
        break;
      case "BYMONTH":
        opts.byMonth = parseByMonthArray(val);
        break;
      case "BYMONTHDAY":
        opts.byMonthDay = parseNumberArray(val);
        break;
      case "BYYEARDAY":
        opts.byYearDay = parseNumberArray(val);
        break;
      case "BYWEEKNO":
        opts.byWeekNo = parseNumberArray(val);
        break;
      case "BYSETPOS":
        opts.bySetPos = parseNumberArray(val);
        break;
      case "WKST":
        opts.wkst = val == null ? void 0 : val.toUpperCase();
        break;
    }
  }
  if (pendingSkip && !opts.rscale) {
    throw new Error("SKIP MUST NOT be present unless RSCALE is present");
  }
  if (pendingSkip && opts.rscale && !opts.skip) {
    opts.skip = pendingSkip;
  }
  return opts;
}
var _RRuleTemporal = class _RRuleTemporal {
  /**
   * Normalize a ZonedDateTime to the polyfill implementation.
   * This prevents type mismatches when mixing native and polyfill Temporal objects.
   */
  static normalizeToPolyfill(zdt) {
    return import_polyfill.Temporal.ZonedDateTime.from(zdt.toString());
  }
  constructor(params) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k;
    let manual;
    if (isIcsOpts(params)) {
      const parsed = parseRRuleString(params.rruleString, params.tzid, params.dtstart, (_a = params.strict) != null ? _a : false);
      if (!parsed.dtstart) {
        throw new Error("dtstart is required - provide it either in rruleString or as a separate parameter");
      }
      this.tzid = (_c = (_b = parsed.tzid) != null ? _b : params.tzid) != null ? _c : "UTC";
      this.originalDtstart = _RRuleTemporal.normalizeToPolyfill(parsed.dtstart);
      manual = __spreadProps(__spreadValues({}, parsed), {
        // Allow explicit COUNT/UNTIL overrides when omitted from the RRULE string
        count: (_d = params.count) != null ? _d : parsed.count,
        until: (_e = params.until) != null ? _e : parsed.until,
        strict: params.strict,
        maxIterations: params.maxIterations,
        includeDtstart: params.includeDtstart,
        tzid: this.tzid
      });
    } else {
      manual = __spreadValues({}, params);
      if (typeof manual.dtstart === "string") {
        throw new Error("Manual dtstart must be a ZonedDateTime");
      }
      manual.tzid = manual.tzid || manual.dtstart.timeZoneId;
      this.tzid = manual.tzid;
      this.originalDtstart = _RRuleTemporal.normalizeToPolyfill(manual.dtstart);
    }
    if (!manual.freq) throw new Error("RRULE must include FREQ");
    manual.interval = (_f = manual.interval) != null ? _f : 1;
    if (manual.interval <= 0) {
      throw new Error("Cannot create RRule: interval must be greater than 0");
    }
    if (manual.until && !(manual.until instanceof import_polyfill.Temporal.ZonedDateTime)) {
      throw new Error("Manual until must be a ZonedDateTime");
    }
    if (manual.until) {
      manual.until = _RRuleTemporal.normalizeToPolyfill(manual.until);
    }
    this.opts = this.sanitizeOpts(manual);
    this.maxIterations = (_g = manual.maxIterations) != null ? _g : 1e4;
    this.includeDtstart = (_h = manual.includeDtstart) != null ? _h : false;
    this.parsedByDayTokens = this.buildParsedByDayTokens(this.opts.byDay);
    this.simpleByDayIsoDays = this.buildByDayIsoDays(this.parsedByDayTokens, false);
    this.allByDayIsoDays = this.buildByDayIsoDays(this.parsedByDayTokens, true);
    this.hasOrdinalByDay = (_j = (_i = this.parsedByDayTokens) == null ? void 0 : _i.some((token) => token.ord !== 0)) != null ? _j : false;
    this.canUseEpochMillisecondsPrecisionFlag = this.originalDtstart.microsecond === 0 && this.originalDtstart.nanosecond === 0 && (!this.opts.until || this.opts.until.microsecond === 0 && this.opts.until.nanosecond === 0);
    this.timeSlotOffsetsMs = this.buildTimeSlotOffsetsMs();
    this.numericByMonths = (_k = this.opts.byMonth) == null ? void 0 : _k.filter((value) => typeof value === "number");
  }
  buildParsedByDayTokens(byDay) {
    if (!(byDay == null ? void 0 : byDay.length)) return void 0;
    const tokens = byDay.map((tok) => {
      const parsed = parseByDayToken(tok);
      if (!parsed) return null;
      return {
        ord: parsed.ord,
        weekday: parsed.weekday,
        isoDay: weekdayToIsoDay[parsed.weekday]
      };
    }).filter((token) => token !== null);
    return tokens.length > 0 ? tokens : void 0;
  }
  buildByDayIsoDays(tokens, includeOrdinals) {
    if (!(tokens == null ? void 0 : tokens.length)) return void 0;
    const isoDays = tokens.filter((token) => includeOrdinals || token.ord === 0).map((token) => token.isoDay);
    if (!isoDays.length) return void 0;
    return [...new Set(isoDays)].sort((a, b) => a - b);
  }
  sanitizeNumericArray(arr, min, max, allowZero = false, sort = false) {
    if (!arr) return void 0;
    const sanitized = arr.filter((n) => Number.isInteger(n) && n >= min && n <= max && (allowZero || n !== 0));
    if (sanitized.length === 0) return void 0;
    return sort ? sanitized.sort((a, b) => a - b) : sanitized;
  }
  sanitizeByDay(byDay) {
    const days = (byDay != null ? byDay : []).filter((day) => Boolean(day) && typeof day === "string");
    const normalized = [];
    for (const day of days) {
      const token = day.toUpperCase();
      const parsed = parseByDayToken(token);
      if (!parsed) {
        throw new Error(`Invalid BYDAY value: ${day}`);
      }
      if (parsed.ord === 0 && /^[+-]?\d/.test(token)) {
        throw new Error(`Invalid BYDAY value: ${day}`);
      }
      normalized.push(token);
    }
    return normalized.length > 0 ? normalized : void 0;
  }
  enforceStrictRfc(opts) {
    var _a;
    if (!opts.strict) return;
    const freq = opts.freq;
    if (opts.byWeekNo && freq !== "YEARLY") {
      throw new Error("BYWEEKNO MUST NOT be used unless FREQ=YEARLY");
    }
    if (opts.byYearDay && ["DAILY", "WEEKLY", "MONTHLY"].includes(freq)) {
      throw new Error("BYYEARDAY MUST NOT be used when FREQ is DAILY, WEEKLY, or MONTHLY");
    }
    if (opts.byMonthDay && freq === "WEEKLY") {
      throw new Error("BYMONTHDAY MUST NOT be used when FREQ is WEEKLY");
    }
    const hasNumericByDay = ((_a = opts.byDay) != null ? _a : []).some((day) => /^[+-]?\d/.test(day));
    if (hasNumericByDay && !["MONTHLY", "YEARLY"].includes(freq)) {
      throw new Error("BYDAY with numeric value MUST NOT be used unless FREQ is MONTHLY or YEARLY");
    }
    if (hasNumericByDay && freq === "YEARLY" && opts.byWeekNo) {
      throw new Error("BYDAY with numeric value MUST NOT be used with FREQ=YEARLY when BYWEEKNO is present");
    }
    const hasOtherBy = Boolean(
      opts.byDay || opts.byMonth || opts.byMonthDay || opts.byYearDay || opts.byWeekNo || opts.byHour || opts.byMinute || opts.bySecond
    );
    if (opts.bySetPos && !hasOtherBy) {
      throw new Error("BYSETPOS MUST be used with another BYxxx rule part");
    }
  }
  sanitizeOpts(opts) {
    var _a;
    if (!allowedFreqSet.has(opts.freq)) {
      throw new Error(`Invalid FREQ value: ${opts.freq}`);
    }
    opts.byDay = this.sanitizeByDay(opts.byDay);
    if (opts.wkst) {
      const wkst = opts.wkst.toUpperCase();
      if (!allowedWeekdaysSet.has(wkst)) {
        throw new Error(`Invalid WKST value: ${opts.wkst}`);
      }
      opts.wkst = wkst;
    }
    if (opts.byMonth) {
      const numeric = opts.byMonth.filter((v) => typeof v === "number");
      const stringy = opts.byMonth.filter((v) => typeof v === "string");
      const sanitizedNum = (_a = this.sanitizeNumericArray(numeric, 1, 12, false, false)) != null ? _a : [];
      const merged = [...sanitizedNum, ...stringy];
      opts.byMonth = merged.length > 0 ? merged : void 0;
    }
    if (opts.rscale && !opts.skip) {
      opts.skip = "OMIT";
    }
    opts.byMonthDay = this.sanitizeNumericArray(opts.byMonthDay, -31, 31, false, false);
    opts.byYearDay = this.sanitizeNumericArray(opts.byYearDay, -366, 366, false, false);
    opts.byWeekNo = this.sanitizeNumericArray(opts.byWeekNo, -53, 53, false, false);
    opts.byHour = this.sanitizeNumericArray(opts.byHour, 0, 23, true, true);
    opts.byMinute = this.sanitizeNumericArray(opts.byMinute, 0, 59, true, true);
    opts.bySecond = this.sanitizeNumericArray(opts.bySecond, 0, 59, true, true);
    if (opts.bySetPos) {
      if (opts.bySetPos.some((p) => p === 0)) {
        throw new Error("bySetPos may not contain 0");
      }
      opts.bySetPos = this.sanitizeNumericArray(opts.bySetPos, -Infinity, Infinity, false, false);
    }
    this.enforceStrictRfc(opts);
    return opts;
  }
  rawAdvance(zdt) {
    const { freq, interval } = this.opts;
    switch (freq) {
      case "DAILY":
        return zdt.add({ days: interval });
      case "WEEKLY":
        return zdt.add({ weeks: interval });
      case "MONTHLY":
        return zdt.add({ months: interval });
      case "YEARLY":
        return zdt.add({ years: interval });
      case "HOURLY": {
        const originalHour = zdt.hour;
        let next = zdt.add({ hours: interval });
        if (next.hour === originalHour && interval === 1) {
          next = next.add({ hours: interval });
        }
        return next;
      }
      case "MINUTELY":
        return zdt.add({ minutes: interval });
      case "SECONDLY":
        return zdt.add({ seconds: interval });
      default:
        throw new Error(`Unsupported FREQ: ${freq}`);
    }
  }
  /**  Expand one base ZonedDateTime into all BYHOUR × BYMINUTE × BYSECOND
   *  combinations, keeping chronological order. If the options are not
   *  present the original date is returned unchanged.
   */
  expandByTime(base) {
    var _a, _b, _c;
    if (!this.opts.byHour && !this.opts.byMinute && !this.opts.bySecond) {
      return [base];
    }
    const hours = (_a = this.opts.byHour) != null ? _a : [base.hour];
    const minutes = (_b = this.opts.byMinute) != null ? _b : [base.minute];
    const seconds = (_c = this.opts.bySecond) != null ? _c : [base.second];
    if (hours.length === 1 && minutes.length === 1 && seconds.length === 1) {
      const hour = hours[0];
      const minute = minutes[0];
      const second = seconds[0];
      if (hour === base.hour && minute === base.minute && second === base.second) {
        return [base];
      }
      return [base.with({ hour, minute, second })];
    }
    const out = [];
    for (const h of hours) {
      for (const m of minutes) {
        for (const s of seconds) {
          out.push(base.with({ hour: h, minute: m, second: s }));
        }
      }
    }
    return out.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
  }
  nextCandidateSameDate(zdt) {
    const { freq, interval = 1, byHour, byMinute, bySecond } = this.opts;
    if (freq === "HOURLY" && byHour && byHour.length === 1) {
      return this.applyTimeOverride(zdt.add({ days: interval }));
    }
    if (freq === "MINUTELY" && byMinute && byMinute.length === 1) {
      return this.applyTimeOverride(zdt.add({ hours: interval }));
    }
    if (bySecond && bySecond.length > 1) {
      const idx = bySecond.indexOf(zdt.second);
      if (idx !== -1 && idx < bySecond.length - 1) {
        return zdt.with({ second: bySecond[idx + 1] });
      }
    }
    if (freq === "MINUTELY" && byHour && byHour.length > 1 && !byMinute) {
      const next = zdt.add({ minutes: interval });
      if (byHour.includes(next.hour)) {
        return next.with({ second: bySecond ? bySecond[0] : zdt.second });
      }
      const nextHour = byHour.find((h) => h > zdt.hour) || byHour[0];
      if (nextHour && nextHour > zdt.hour) {
        return zdt.with({ hour: nextHour, minute: 0, second: bySecond ? bySecond[0] : zdt.second });
      }
      return this.applyTimeOverride(zdt.add({ days: 1 }));
    }
    if (freq === "SECONDLY") {
      let candidate = zdt;
      if (bySecond && bySecond.length > 0) {
        const nextSecondInList = bySecond.find((s) => s > candidate.second);
        if (nextSecondInList !== void 0) {
          return candidate.with({ second: nextSecondInList });
        }
        candidate = candidate.with({ second: bySecond[0] }).add({ minutes: 1 });
      } else {
        candidate = candidate.add({ seconds: interval });
      }
      if (byMinute && byMinute.length > 0) {
        if (!byMinute.includes(candidate.minute) || candidate.minute === zdt.minute && candidate.second < zdt.second) {
          const nextMinuteInList = byMinute.find((m) => m > candidate.minute);
          if (nextMinuteInList !== void 0) {
            return candidate.with({ minute: nextMinuteInList, second: bySecond ? bySecond[0] : 0 });
          }
          candidate = candidate.with({ minute: byMinute[0], second: bySecond ? bySecond[0] : 0 }).add({ hours: 1 });
        }
      }
      if (byHour && byHour.length > 0) {
        if (!byHour.includes(candidate.hour) || candidate.hour === zdt.hour && candidate.minute < zdt.minute) {
          const nextHourInList = byHour.find((h) => h > candidate.hour);
          if (nextHourInList !== void 0) {
            return candidate.with({
              hour: nextHourInList,
              minute: byMinute ? byMinute[0] : 0,
              second: bySecond ? bySecond[0] : 0
            });
          }
          candidate = candidate.with({ hour: byHour[0], minute: byMinute ? byMinute[0] : 0, second: bySecond ? bySecond[0] : 0 }).add({ days: 1 });
        }
      }
      return candidate;
    }
    if (byMinute && byMinute.length > 1) {
      const idx = byMinute.indexOf(zdt.minute);
      if (idx !== -1 && idx < byMinute.length - 1) {
        return zdt.with({
          minute: byMinute[idx + 1],
          second: bySecond ? bySecond[0] : zdt.second
        });
      }
      if (freq === "MINUTELY" && idx === byMinute.length - 1) {
        if (byHour && byHour.length > 0) {
          const currentHourIdx = byHour.indexOf(zdt.hour);
          if (currentHourIdx !== -1 && currentHourIdx < byHour.length - 1) {
            return zdt.with({
              hour: byHour[currentHourIdx + 1],
              minute: byMinute[0],
              second: bySecond ? bySecond[0] : zdt.second
            });
          } else {
            return this.applyTimeOverride(zdt.add({ days: 1 }));
          }
        }
        return zdt.add({ hours: interval }).with({
          minute: byMinute[0],
          second: bySecond ? bySecond[0] : zdt.second
        });
      }
    }
    if (byHour && byHour.length > 1) {
      const idx = byHour.indexOf(zdt.hour);
      if (idx !== -1 && idx < byHour.length - 1) {
        return zdt.with({
          hour: byHour[idx + 1],
          minute: byMinute ? byMinute[0] : zdt.minute,
          second: bySecond ? bySecond[0] : zdt.second
        });
      }
    }
    if (freq === "HOURLY" && byHour && byHour.length > 1) {
      return this.applyTimeOverride(zdt.add({ days: 1 }));
    }
    return this.applyTimeOverride(this.rawAdvance(zdt));
  }
  applyTimeOverride(zdt) {
    const { byHour, byMinute, bySecond } = this.opts;
    let dt = zdt;
    if (byHour) dt = dt.with({ hour: byHour[0] });
    if (byMinute) dt = dt.with({ minute: byMinute[0] });
    if (bySecond) dt = dt.with({ second: bySecond[0] });
    return dt;
  }
  computeFirst() {
    var _a, _b, _c, _d;
    let zdt = this.originalDtstart;
    if (((_a = this.opts.byWeekNo) == null ? void 0 : _a.length) && ["DAILY", "HOURLY", "MINUTELY", "SECONDLY"].includes(this.opts.freq)) {
      let targetWeek = this.opts.byWeekNo[0];
      let targetYear = zdt.year;
      while (targetYear <= zdt.year + 10) {
        const jan1 = zdt.with({ year: targetYear, month: 1, day: 1 });
        const dec31 = zdt.with({ year: targetYear, month: 12, day: 31 });
        let hasTargetWeek = false;
        if (targetWeek > 0) {
          let maxWeek = 52;
          if (jan1.dayOfWeek === 4 || dec31.dayOfWeek === 4) {
            maxWeek = 53;
          }
          hasTargetWeek = targetWeek <= maxWeek;
        } else {
          let maxWeek = 52;
          if (jan1.dayOfWeek === 4 || dec31.dayOfWeek === 4) {
            maxWeek = 53;
          }
          hasTargetWeek = -targetWeek <= maxWeek;
        }
        if (hasTargetWeek) {
          const firstThursday = jan1.add({ days: (4 - jan1.dayOfWeek + 7) % 7 });
          let weekStart;
          if (targetWeek > 0) {
            weekStart = firstThursday.subtract({ days: 3 }).add({ weeks: targetWeek - 1 });
          } else {
            const lastWeek = jan1.dayOfWeek === 4 || dec31.dayOfWeek === 4 ? 53 : 52;
            weekStart = firstThursday.subtract({ days: 3 }).add({ weeks: lastWeek + targetWeek });
          }
          if ((_b = this.opts.byDay) == null ? void 0 : _b.length) {
            const dayMap = weekdayToIsoDay;
            const targetDays = this.opts.byDay.map((tok) => extractWeekdayToken(tok)).filter((day) => day !== null).map((day) => dayMap[day]).filter(Boolean);
            if (targetDays.length) {
              const candidates = targetDays.map((dayOfWeek) => {
                const delta = (dayOfWeek - weekStart.dayOfWeek + 7) % 7;
                return weekStart.add({ days: delta });
              });
              const firstCandidate = candidates.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b))[0];
              if (firstCandidate && import_polyfill.Temporal.ZonedDateTime.compare(firstCandidate, this.originalDtstart) >= 0) {
                zdt = firstCandidate;
                break;
              }
            }
          } else {
            if (import_polyfill.Temporal.ZonedDateTime.compare(weekStart, this.originalDtstart) >= 0) {
              zdt = weekStart;
              break;
            }
          }
        }
        targetYear++;
      }
    }
    if (((_c = this.opts.byDay) == null ? void 0 : _c.length) && !this.opts.byWeekNo) {
      const dayMap = weekdayToIsoDay;
      const hasOrdinalTokens = this.opts.byDay.some((tok) => /^[+-]?\d/.test(tok));
      if (hasOrdinalTokens && this.opts.byMonth && (this.opts.freq === "MINUTELY" || this.opts.freq === "SECONDLY")) {
        const months = this.opts.byMonth.filter((v) => typeof v === "number").sort((a, b) => a - b);
        let foundFirst = false;
        for (let year = zdt.year; year <= zdt.year + 10 && !foundFirst; year++) {
          for (const month of months) {
            if (year === zdt.year && month < zdt.month) continue;
            const monthSample = zdt.with({ year, month, day: 1 });
            const monthlyOccs = this.generateMonthlyOccurrences(monthSample);
            for (const occ of monthlyOccs) {
              if (import_polyfill.Temporal.ZonedDateTime.compare(occ, zdt) >= 0) {
                if (!occ.toPlainDate().equals(zdt.toPlainDate())) {
                  zdt = this.applyTimeOverride(occ.with({ hour: 0, minute: 0, second: 0 }));
                } else {
                  zdt = occ;
                }
                foundFirst = true;
                break;
              }
            }
            if (foundFirst) break;
          }
        }
      } else {
        let deltas;
        const weekdayTokens = this.opts.byDay.map((tok) => extractWeekdayToken(tok)).filter((tok) => tok !== null);
        if (["DAILY", "HOURLY", "MINUTELY", "SECONDLY"].includes(this.opts.freq) && weekdayTokens.length === this.opts.byDay.length) {
          deltas = weekdayTokens.map((tok) => (dayMap[tok] - zdt.dayOfWeek + 7) % 7);
        } else {
          deltas = weekdayTokens.map((wdTok) => (dayMap[wdTok] - zdt.dayOfWeek + 7) % 7);
        }
        if (deltas.length) {
          zdt = zdt.add({ days: Math.min(...deltas) });
        }
      }
    }
    const { byHour, byMinute, bySecond } = this.opts;
    if (this.opts.freq === "HOURLY" && !byHour && import_polyfill.Temporal.ZonedDateTime.compare(
      zdt.with({ hour: 0, minute: 0, second: 0, microsecond: 0, nanosecond: 0 }),
      this.originalDtstart
    ) > 0) {
      zdt = zdt.with({ hour: 0, minute: 0, second: 0, microsecond: 0, nanosecond: 0 });
    }
    if (this.opts.freq === "MINUTELY" && !byMinute && import_polyfill.Temporal.ZonedDateTime.compare(
      zdt.with({ hour: 0, minute: 0, second: 0, microsecond: 0, nanosecond: 0 }),
      this.originalDtstart
    ) > 0) {
      zdt = zdt.with({ hour: 0, minute: 0, second: 0, microsecond: 0, nanosecond: 0 });
    }
    if (this.opts.freq === "SECONDLY" && ((_d = this.opts.byWeekNo) == null ? void 0 : _d.length) && !bySecond && import_polyfill.Temporal.ZonedDateTime.compare(
      zdt.with({ hour: 0, minute: 0, second: 0, microsecond: 0, nanosecond: 0 }),
      this.originalDtstart
    ) > 0) {
      zdt = zdt.with({ hour: 0, minute: 0, second: 0, microsecond: 0, nanosecond: 0 });
    }
    if (byHour || byMinute || bySecond) {
      const candidates = this.expandByTime(zdt);
      for (const candidate of candidates) {
        if (import_polyfill.Temporal.ZonedDateTime.compare(candidate, this.originalDtstart) >= 0) {
          return candidate;
        }
      }
      zdt = this.applyTimeOverride(this.rawAdvance(zdt));
    }
    return zdt;
  }
  // --- NEW: constraint checks ---
  // 2) Replace your matchesByDay with this:
  matchesByDay(zdt) {
    var _a, _b, _c;
    const { byDay, freq } = this.opts;
    if (!byDay) return true;
    if (!this.hasOrdinalByDay) {
      return (_b = (_a = this.simpleByDayIsoDays) == null ? void 0 : _a.includes(zdt.dayOfWeek)) != null ? _b : false;
    }
    for (const token of (_c = this.parsedByDayTokens) != null ? _c : []) {
      if (freq === "DAILY" && zdt.dayOfWeek === token.isoDay) return true;
      if (token.ord === 0) {
        if (zdt.dayOfWeek === token.isoDay) return true;
        continue;
      }
      const month = zdt.month;
      let dt = zdt.with({ day: 1 });
      const candidates = [];
      while (dt.month === month) {
        if (dt.dayOfWeek === token.isoDay) candidates.push(dt.day);
        dt = dt.add({ days: 1 });
      }
      const idx = token.ord > 0 ? token.ord - 1 : candidates.length + token.ord;
      if (candidates[idx] === zdt.day) return true;
    }
    return false;
  }
  matchesByMonth(zdt) {
    const { byMonth } = this.opts;
    if (!byMonth) return true;
    const nums = byMonth.filter((v) => typeof v === "number");
    if (nums.length === 0) return true;
    return nums.includes(zdt.month);
  }
  matchesNumericConstraint(value, constraints, maxPositiveValue) {
    return constraints.some((c) => {
      const target = c > 0 ? c : maxPositiveValue + c + 1;
      return value === target;
    });
  }
  matchesByMonthDay(zdt) {
    const { byMonthDay } = this.opts;
    if (!byMonthDay) return true;
    const lastDay = zdt.with({ day: 1 }).add({ months: 1 }).subtract({ days: 1 }).day;
    return this.matchesNumericConstraint(zdt.day, byMonthDay, lastDay);
  }
  matchesByHour(zdt) {
    const { byHour } = this.opts;
    if (!byHour) return true;
    if (byHour.includes(zdt.hour)) {
      return true;
    }
    for (const h of byHour) {
      const intendedTime = zdt.with({ hour: h });
      if (intendedTime.hour === zdt.hour) {
        return true;
      }
    }
    return false;
  }
  matchesByMinute(zdt) {
    const { byMinute } = this.opts;
    if (!byMinute) return true;
    return byMinute.includes(zdt.minute);
  }
  matchesBySecond(zdt) {
    const { bySecond } = this.opts;
    if (!bySecond) return true;
    return bySecond.includes(zdt.second);
  }
  matchesAll(zdt) {
    return this.matchesByMonth(zdt) && this.matchesByWeekNo(zdt) && this.matchesByYearDay(zdt) && this.matchesByMonthDay(zdt) && this.matchesByDay(zdt) && this.matchesByHour(zdt) && this.matchesByMinute(zdt) && this.matchesBySecond(zdt);
  }
  matchesByYearDay(zdt) {
    const { byYearDay } = this.opts;
    if (!byYearDay) return true;
    const dayOfYear = zdt.dayOfYear;
    const last = zdt.with({ month: 12, day: 31 }).dayOfYear;
    return this.matchesNumericConstraint(dayOfYear, byYearDay, last);
  }
  getIsoWeekInfo(zdt) {
    const thursday = zdt.add({ days: 4 - zdt.dayOfWeek });
    const year = thursday.year;
    const jan1 = zdt.with({ year, month: 1, day: 1 });
    const firstThursday = jan1.add({ days: (4 - jan1.dayOfWeek + 7) % 7 });
    const diffDays = thursday.toPlainDate().since(firstThursday.toPlainDate()).days;
    const week = Math.floor(diffDays / 7) + 1;
    return { week, year };
  }
  matchesByWeekNo(zdt) {
    const { byWeekNo } = this.opts;
    if (!byWeekNo) return true;
    const { week, year } = this.getIsoWeekInfo(zdt);
    const jan1 = zdt.with({ year, month: 1, day: 1 });
    const isLeapYear = jan1.inLeapYear;
    const lastWeek = jan1.dayOfWeek === 4 || isLeapYear && jan1.dayOfWeek === 3 ? 53 : 52;
    return byWeekNo.some((wn) => {
      if (wn > 0) {
        return week === wn;
      } else {
        return week === lastWeek + wn + 1;
      }
    });
  }
  options() {
    return this.opts;
  }
  cloneOptions() {
    const _a = this.opts, {
      byHour,
      byMinute,
      bySecond,
      byDay,
      byMonth,
      byMonthDay,
      byYearDay,
      byWeekNo,
      bySetPos,
      rDate,
      exDate
    } = _a, rest = __objRest(_a, [
      "byHour",
      "byMinute",
      "bySecond",
      "byDay",
      "byMonth",
      "byMonthDay",
      "byYearDay",
      "byWeekNo",
      "bySetPos",
      "rDate",
      "exDate"
    ]);
    return __spreadProps(__spreadValues({}, rest), {
      byHour: byHour ? [...byHour] : void 0,
      byMinute: byMinute ? [...byMinute] : void 0,
      bySecond: bySecond ? [...bySecond] : void 0,
      byDay: byDay ? [...byDay] : void 0,
      byMonth: byMonth ? [...byMonth] : void 0,
      byMonthDay: byMonthDay ? [...byMonthDay] : void 0,
      byYearDay: byYearDay ? [...byYearDay] : void 0,
      byWeekNo: byWeekNo ? [...byWeekNo] : void 0,
      bySetPos: bySetPos ? [...bySetPos] : void 0,
      rDate: rDate ? [...rDate] : void 0,
      exDate: exDate ? [...exDate] : void 0
    });
  }
  cloneUpdateOptions(updates) {
    const cloned = {};
    if (Object.prototype.hasOwnProperty.call(updates, "byHour")) {
      cloned.byHour = Array.isArray(updates.byHour) ? [...updates.byHour] : updates.byHour;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "byMinute")) {
      cloned.byMinute = Array.isArray(updates.byMinute) ? [...updates.byMinute] : updates.byMinute;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "bySecond")) {
      cloned.bySecond = Array.isArray(updates.bySecond) ? [...updates.bySecond] : updates.bySecond;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "byDay")) {
      cloned.byDay = Array.isArray(updates.byDay) ? [...updates.byDay] : updates.byDay;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "byMonth")) {
      cloned.byMonth = Array.isArray(updates.byMonth) ? [...updates.byMonth] : updates.byMonth;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "byMonthDay")) {
      cloned.byMonthDay = Array.isArray(updates.byMonthDay) ? [...updates.byMonthDay] : updates.byMonthDay;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "byYearDay")) {
      cloned.byYearDay = Array.isArray(updates.byYearDay) ? [...updates.byYearDay] : updates.byYearDay;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "byWeekNo")) {
      cloned.byWeekNo = Array.isArray(updates.byWeekNo) ? [...updates.byWeekNo] : updates.byWeekNo;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "bySetPos")) {
      cloned.bySetPos = Array.isArray(updates.bySetPos) ? [...updates.bySetPos] : updates.bySetPos;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "rDate")) {
      cloned.rDate = Array.isArray(updates.rDate) ? [...updates.rDate] : updates.rDate;
    }
    if (Object.prototype.hasOwnProperty.call(updates, "exDate")) {
      cloned.exDate = Array.isArray(updates.exDate) ? [...updates.exDate] : updates.exDate;
    }
    return cloned;
  }
  /**
   * Create a new {@link RRuleTemporal} instance with modified options while keeping the current one unchanged.
   *
   * @example
   * ```ts
   * const updated = rule.with({byMonthDay: [3]});
   * ```
   */
  with(updates) {
    var _a, _b;
    const merged = __spreadProps(__spreadValues(__spreadValues(__spreadValues({}, this.cloneOptions()), updates), this.cloneUpdateOptions(updates)), {
      tzid: (_a = updates.tzid) != null ? _a : this.opts.tzid,
      dtstart: (_b = updates.dtstart) != null ? _b : this.opts.dtstart
    });
    return new _RRuleTemporal(merged);
  }
  addDtstartIfNeeded(dates, iterator) {
    if (this.includeDtstart && !this.matchesAll(this.originalDtstart)) {
      if (iterator && this.isExcluded(this.originalDtstart)) {
        return true;
      }
      if (iterator && !iterator(this.originalDtstart, dates.length)) {
        return false;
      }
      dates.push(this.originalDtstart);
      if (this.shouldBreakForCountLimit(dates.length)) {
        return false;
      }
    }
    return true;
  }
  canUseUtcLinearFastPath(iterator) {
    if (iterator || this.tzid !== "UTC" || this.opts.rscale || this.opts.rDate || this.opts.exDate) {
      return false;
    }
    if (this.opts.byMonth || this.opts.byMonthDay || this.opts.byYearDay || this.opts.byWeekNo || this.opts.bySetPos) {
      return false;
    }
    switch (this.opts.freq) {
      case "DAILY":
        return !this.opts.byHour && !this.opts.byMinute && !this.opts.bySecond && !this.hasOrdinalByDay;
      case "HOURLY":
      case "MINUTELY":
        return !this.opts.byDay && !this.opts.byHour && !this.opts.byMinute && !this.opts.bySecond;
      default:
        return false;
    }
  }
  canUseUtcWeeklyFastPath(iterator) {
    return !iterator && this.tzid === "UTC" && this.opts.freq === "WEEKLY" && !this.opts.rscale && !this.opts.rDate && !this.opts.exDate && !this.opts.byMonth && !this.opts.byMonthDay && !this.opts.byYearDay && !this.opts.byWeekNo && !this.opts.bySetPos && !this.opts.byHour && !this.opts.byMinute && !this.opts.bySecond && !this.hasOrdinalByDay;
  }
  canUseUtcMonthlyFastPath(iterator) {
    return !iterator && this.tzid === "UTC" && this.opts.freq === "MONTHLY" && !this.opts.rscale && !this.opts.rDate && !this.opts.exDate && !this.opts.byYearDay && !this.opts.byWeekNo && this.canUseEpochMillisecondsPrecisionFlag && !!(this.opts.byDay || this.opts.byMonthDay);
  }
  utcZdtFromEpochNanoseconds(epochNanoseconds) {
    return import_polyfill.Temporal.Instant.fromEpochNanoseconds(epochNanoseconds).toZonedDateTimeISO("UTC");
  }
  utcZdtFromEpochMilliseconds(epochMilliseconds) {
    return import_polyfill.Temporal.Instant.fromEpochMilliseconds(epochMilliseconds).toZonedDateTimeISO("UTC");
  }
  canUseUtcEpochMillisecondsPrecision() {
    return this.canUseEpochMillisecondsPrecisionFlag;
  }
  buildTimeSlotOffsetsMs() {
    var _a, _b, _c;
    if (!this.canUseEpochMillisecondsPrecisionFlag) return void 0;
    const hours = (_a = this.opts.byHour) != null ? _a : [this.originalDtstart.hour];
    const minutes = (_b = this.opts.byMinute) != null ? _b : [this.originalDtstart.minute];
    const seconds = (_c = this.opts.bySecond) != null ? _c : [this.originalDtstart.second];
    const baseMilliseconds = this.originalDtstart.millisecond;
    const offsets = [];
    for (const hour of hours) {
      for (const minute of minutes) {
        for (const second of seconds) {
          offsets.push(((hour * 60 + minute) * 60 + second) * MS_PER_SECOND + baseMilliseconds);
        }
      }
    }
    return offsets;
  }
  findFirstMatchingDailyStep(startDayOfWeek, stepDays, allowedDays) {
    let dayOfWeek = startDayOfWeek;
    for (let steps = 0; steps < 7; steps++) {
      if (allowedDays.includes(dayOfWeek)) {
        return steps;
      }
      dayOfWeek = addIsoDays(dayOfWeek, stepDays);
    }
    return null;
  }
  allUtcFastPath(iterator) {
    if (this.canUseUtcLinearFastPath(iterator)) {
      switch (this.opts.freq) {
        case "DAILY":
          return this._allUtcDailySimple();
        case "HOURLY":
          return this._allUtcFixedStepSimple(NS_PER_HOUR * BigInt(this.opts.interval));
        case "MINUTELY":
          return this._allUtcFixedStepSimple(NS_PER_MINUTE * BigInt(this.opts.interval));
      }
    }
    if (this.canUseUtcMonthlyFastPath(iterator)) {
      return this._allUtcMonthlyByDayOrMonthDay();
    }
    if (this.canUseUtcWeeklyFastPath(iterator)) {
      return this._allUtcWeeklySimple();
    }
    return null;
  }
  _allUtcFixedStepSimple(stepNanoseconds) {
    var _a, _b;
    const dates = [];
    if (!this.addDtstartIfNeeded(dates)) {
      return dates;
    }
    let iterationCount = 0;
    if (this.canUseUtcEpochMillisecondsPrecision()) {
      let currentMilliseconds = this.originalDtstart.epochMilliseconds;
      const stepMilliseconds = Number(stepNanoseconds / NS_PER_MILLISECOND);
      const untilMilliseconds = (_a = this.opts.until) == null ? void 0 : _a.epochMilliseconds;
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        if (untilMilliseconds !== void 0 && currentMilliseconds > untilMilliseconds) {
          break;
        }
        dates.push(this.utcZdtFromEpochMilliseconds(currentMilliseconds));
        if (this.shouldBreakForCountLimit(dates.length)) {
          break;
        }
        currentMilliseconds += stepMilliseconds;
      }
      return dates;
    }
    let currentNanoseconds = this.originalDtstart.epochNanoseconds;
    const untilNanoseconds = (_b = this.opts.until) == null ? void 0 : _b.epochNanoseconds;
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      if (untilNanoseconds !== void 0 && currentNanoseconds > untilNanoseconds) {
        break;
      }
      dates.push(this.utcZdtFromEpochNanoseconds(currentNanoseconds));
      if (this.shouldBreakForCountLimit(dates.length)) {
        break;
      }
      currentNanoseconds += stepNanoseconds;
    }
    return dates;
  }
  _allUtcDailySimple() {
    var _a, _b;
    const dates = [];
    if (!this.addDtstartIfNeeded(dates)) {
      return dates;
    }
    const stepDays = this.opts.interval;
    const allowedDays = this.simpleByDayIsoDays;
    let iterationCount = 0;
    if (this.canUseUtcEpochMillisecondsPrecision()) {
      const stepMilliseconds = stepDays * MS_PER_DAY;
      const untilMilliseconds = (_a = this.opts.until) == null ? void 0 : _a.epochMilliseconds;
      let currentMilliseconds = this.originalDtstart.epochMilliseconds;
      let currentDayOfWeek2 = this.originalDtstart.dayOfWeek;
      if (allowedDays == null ? void 0 : allowedDays.length) {
        const firstMatchingStep = this.findFirstMatchingDailyStep(currentDayOfWeek2, stepDays, allowedDays);
        if (firstMatchingStep === null) {
          return dates;
        }
        currentMilliseconds += firstMatchingStep * stepMilliseconds;
        currentDayOfWeek2 = addIsoDays(currentDayOfWeek2, firstMatchingStep * stepDays);
      }
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        if (untilMilliseconds !== void 0 && currentMilliseconds > untilMilliseconds) {
          break;
        }
        if (!allowedDays || allowedDays.includes(currentDayOfWeek2)) {
          dates.push(this.utcZdtFromEpochMilliseconds(currentMilliseconds));
          if (this.shouldBreakForCountLimit(dates.length)) {
            break;
          }
        }
        currentMilliseconds += stepMilliseconds;
        currentDayOfWeek2 = addIsoDays(currentDayOfWeek2, stepDays);
      }
      return dates;
    }
    const stepNanoseconds = BigInt(stepDays) * NS_PER_DAY;
    const untilNanoseconds = (_b = this.opts.until) == null ? void 0 : _b.epochNanoseconds;
    let currentNanoseconds = this.originalDtstart.epochNanoseconds;
    let currentDayOfWeek = this.originalDtstart.dayOfWeek;
    if (allowedDays == null ? void 0 : allowedDays.length) {
      const firstMatchingStep = this.findFirstMatchingDailyStep(currentDayOfWeek, stepDays, allowedDays);
      if (firstMatchingStep === null) {
        return dates;
      }
      currentNanoseconds += BigInt(firstMatchingStep) * stepNanoseconds;
      currentDayOfWeek = addIsoDays(currentDayOfWeek, firstMatchingStep * stepDays);
    }
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      if (untilNanoseconds !== void 0 && currentNanoseconds > untilNanoseconds) {
        break;
      }
      if (!allowedDays || allowedDays.includes(currentDayOfWeek)) {
        dates.push(this.utcZdtFromEpochNanoseconds(currentNanoseconds));
        if (this.shouldBreakForCountLimit(dates.length)) {
          break;
        }
      }
      currentNanoseconds += stepNanoseconds;
      currentDayOfWeek = addIsoDays(currentDayOfWeek, stepDays);
    }
    return dates;
  }
  _allUtcWeeklySimple() {
    var _a, _b, _c, _d, _e;
    const dates = [];
    if (!this.addDtstartIfNeeded(dates)) {
      return dates;
    }
    const start = this.originalDtstart;
    const wkstToken = (_a = extractWeekdayToken(this.opts.wkst || "MO")) != null ? _a : "MO";
    const wkstDay = (_b = weekdayToIsoDay[wkstToken]) != null ? _b : 1;
    const targetDays = this.opts.byDay ? [...(_c = this.allByDayIsoDays) != null ? _c : []] : [start.dayOfWeek];
    const dayOffsets = targetDays.map((day) => (day - wkstDay + 7) % 7).sort((a, b) => a - b);
    const weekStartOffset = (start.dayOfWeek - wkstDay + 7) % 7;
    let iterationCount = 0;
    if (this.canUseUtcEpochMillisecondsPrecision()) {
      const startMilliseconds = start.epochMilliseconds;
      const untilMilliseconds = (_d = this.opts.until) == null ? void 0 : _d.epochMilliseconds;
      const weekStepMilliseconds = this.opts.interval * MS_PER_WEEK;
      let weekStartMilliseconds = startMilliseconds - weekStartOffset * MS_PER_DAY;
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        for (const dayOffset of dayOffsets) {
          const occurrenceMilliseconds = weekStartMilliseconds + dayOffset * MS_PER_DAY;
          if (occurrenceMilliseconds < startMilliseconds) {
            continue;
          }
          if (untilMilliseconds !== void 0 && occurrenceMilliseconds > untilMilliseconds) {
            return dates;
          }
          dates.push(this.utcZdtFromEpochMilliseconds(occurrenceMilliseconds));
          if (this.shouldBreakForCountLimit(dates.length)) {
            return dates;
          }
        }
        weekStartMilliseconds += weekStepMilliseconds;
      }
    }
    const startNanoseconds = start.epochNanoseconds;
    const untilNanoseconds = (_e = this.opts.until) == null ? void 0 : _e.epochNanoseconds;
    const weekStepNanoseconds = BigInt(this.opts.interval) * NS_PER_WEEK;
    let weekStartNanoseconds = startNanoseconds - BigInt(weekStartOffset) * NS_PER_DAY;
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      for (const dayOffset of dayOffsets) {
        const occurrenceNanoseconds = weekStartNanoseconds + BigInt(dayOffset) * NS_PER_DAY;
        if (occurrenceNanoseconds < startNanoseconds) {
          continue;
        }
        if (untilNanoseconds !== void 0 && occurrenceNanoseconds > untilNanoseconds) {
          return dates;
        }
        dates.push(this.utcZdtFromEpochNanoseconds(occurrenceNanoseconds));
        if (this.shouldBreakForCountLimit(dates.length)) {
          return dates;
        }
      }
      weekStartNanoseconds += weekStepNanoseconds;
    }
  }
  hasSingleExpandedTimeSlot() {
    var _a, _b, _c;
    if (this.timeSlotOffsetsMs) {
      return this.timeSlotOffsetsMs.length === 1;
    }
    const hours = (_a = this.opts.byHour) != null ? _a : [this.originalDtstart.hour];
    const minutes = (_b = this.opts.byMinute) != null ? _b : [this.originalDtstart.minute];
    const seconds = (_c = this.opts.bySecond) != null ? _c : [this.originalDtstart.second];
    return hours.length === 1 && minutes.length === 1 && seconds.length === 1;
  }
  buildMonthlyOccurrenceOnDay(monthStart, day) {
    const base = monthStart.day === day ? monthStart : monthStart.with({ day });
    return this.applyTimeOverride(base);
  }
  applyBySetPosToSortedList(list2) {
    const { bySetPos } = this.opts;
    if (!bySetPos || !bySetPos.length || list2.length === 0) return list2;
    const out = [];
    const len = list2.length;
    for (const pos of bySetPos) {
      const idx = pos > 0 ? pos - 1 : len + pos;
      if (idx >= 0 && idx < len) out.push(list2[idx]);
    }
    return out;
  }
  generateMonthlyOccurrenceDays(sample) {
    const { byDay, byMonth, byMonthDay } = this.opts;
    const monthStart = sample.day === 1 ? sample : sample.with({ day: 1 });
    if (byMonth && !byMonth.includes(sample.month)) return [];
    const lastDay = monthStart.add({ months: 1 }).subtract({ days: 1 }).day;
    let byMonthDayHits = [];
    if (byMonthDay && byMonthDay.length > 0) {
      byMonthDayHits = byMonthDay.map((d) => d > 0 ? d : lastDay + d + 1).filter((d) => d >= 1 && d <= lastDay);
      byMonthDayHits = [...new Set(byMonthDayHits)].sort((a, b) => a - b);
    }
    if (!byDay && byMonthDay && byMonthDay.length > 0) {
      return byMonthDayHits;
    }
    if (!byDay) {
      return [sample.day];
    }
    const tokens = this.parsedByDayTokens;
    if (!(tokens == null ? void 0 : tokens.length)) return [];
    const firstDayOfWeek = monthStart.dayOfWeek;
    const lastDayOfWeek = (firstDayOfWeek - 1 + lastDay - 1) % 7 + 1;
    const byDayHits = /* @__PURE__ */ new Set();
    for (const { ord, isoDay } of tokens) {
      if (ord === 0) {
        let day = 1 + (isoDay - firstDayOfWeek + 7) % 7;
        while (day <= lastDay) {
          byDayHits.add(day);
          day += 7;
        }
      } else {
        let day;
        if (ord > 0) {
          day = 1 + (isoDay - firstDayOfWeek + 7) % 7 + 7 * (ord - 1);
        } else {
          const lastMatch = lastDay - (lastDayOfWeek - isoDay + 7) % 7;
          day = lastMatch + 7 * (ord + 1);
        }
        if (day >= 1 && day <= lastDay) {
          byDayHits.add(day);
        }
      }
    }
    let finalDays = [...byDayHits].sort((a, b) => a - b);
    if (byMonthDay && byMonthDay.length > 0) {
      if (byMonthDayHits.length === 0) {
        return [];
      }
      const byMonthDayHitSet = new Set(byMonthDayHits);
      finalDays = finalDays.filter((d) => byMonthDayHitSet.has(d));
    }
    return finalDays;
  }
  generateMonthlyOccurrencesOptimizedBySetPos(sample) {
    if (!this.opts.bySetPos || !this.hasSingleExpandedTimeSlot()) {
      return null;
    }
    const monthStart = sample.day === 1 ? sample : sample.with({ day: 1 });
    const days = this.generateMonthlyOccurrenceDays(monthStart);
    if (days.length === 0) {
      return [];
    }
    const selectedDays = this.applyBySetPosToSortedList(days);
    if (selectedDays.length === 0) {
      return [];
    }
    return selectedDays.sort((a, b) => a - b).map((day) => this.buildMonthlyOccurrenceOnDay(monthStart, day));
  }
  isGregorianLeapYear(year) {
    return year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
  }
  daysInGregorianMonth(year, month) {
    if (month === 2 && this.isGregorianLeapYear(year)) {
      return 29;
    }
    return GREGORIAN_MONTH_LENGTHS[month - 1];
  }
  gregorianIsoDayOfWeek(year, month, day) {
    let adjustedYear = year;
    if (month < 3) adjustedYear -= 1;
    const sundayZero = (adjustedYear + Math.floor(adjustedYear / 4) - Math.floor(adjustedYear / 100) + Math.floor(adjustedYear / 400) + GREGORIAN_WEEKDAY_OFFSETS[month - 1] + day) % 7;
    return sundayZero === 0 ? 7 : sundayZero;
  }
  monthIndexToYearMonth(monthIndex) {
    const year = Math.floor(monthIndex / 12);
    return {
      year,
      month: monthIndex - year * 12 + 1
    };
  }
  generateMonthlyOccurrenceDaysUtc(year, month) {
    if (this.numericByMonths && this.numericByMonths.length > 0 && !this.numericByMonths.includes(month)) {
      return [];
    }
    const byMonthDay = this.opts.byMonthDay;
    const byDay = this.opts.byDay;
    const lastDay = this.daysInGregorianMonth(year, month);
    let byMonthDayHits = [];
    if (byMonthDay && byMonthDay.length > 0) {
      byMonthDayHits = byMonthDay.map((day) => day > 0 ? day : lastDay + day + 1).filter((day) => day >= 1 && day <= lastDay);
      byMonthDayHits = [...new Set(byMonthDayHits)].sort((a, b) => a - b);
    }
    if (!byDay && byMonthDay && byMonthDay.length > 0) {
      return byMonthDayHits;
    }
    if (!byDay) {
      const day = this.originalDtstart.day;
      return day >= 1 && day <= lastDay ? [day] : [];
    }
    const tokens = this.parsedByDayTokens;
    if (!(tokens == null ? void 0 : tokens.length)) return [];
    const firstDayOfWeek = this.gregorianIsoDayOfWeek(year, month, 1);
    const lastDayOfWeek = addIsoDays(firstDayOfWeek, lastDay - 1);
    const byDayHits = /* @__PURE__ */ new Set();
    for (const { ord, isoDay } of tokens) {
      if (ord === 0) {
        let day = 1 + (isoDay - firstDayOfWeek + 7) % 7;
        while (day <= lastDay) {
          byDayHits.add(day);
          day += 7;
        }
      } else {
        let day;
        if (ord > 0) {
          day = 1 + (isoDay - firstDayOfWeek + 7) % 7 + 7 * (ord - 1);
        } else {
          const lastMatch = lastDay - (lastDayOfWeek - isoDay + 7) % 7;
          day = lastMatch + 7 * (ord + 1);
        }
        if (day >= 1 && day <= lastDay) {
          byDayHits.add(day);
        }
      }
    }
    let finalDays = [...byDayHits].sort((a, b) => a - b);
    if (byMonthDay && byMonthDay.length > 0) {
      if (byMonthDayHits.length === 0) return [];
      const byMonthDayHitSet = new Set(byMonthDayHits);
      finalDays = finalDays.filter((day) => byMonthDayHitSet.has(day));
    }
    return finalDays;
  }
  generateMonthlyOccurrenceEpochsUtc(year, month) {
    var _a;
    const days = this.generateMonthlyOccurrenceDaysUtc(year, month);
    if (days.length === 0) return [];
    const monthStartMs = Date.UTC(year, month - 1, 1, 0, 0, 0, 0);
    const timeSlotOffsets = (_a = this.timeSlotOffsetsMs) != null ? _a : [0];
    if (this.opts.bySetPos && this.opts.bySetPos.length > 0) {
      if (timeSlotOffsets.length === 1) {
        const selectedDays = this.applyBySetPosToSortedList(days).sort((a, b) => a - b);
        const offset = timeSlotOffsets[0];
        return selectedDays.map((day) => monthStartMs + (day - 1) * MS_PER_DAY + offset);
      }
      const timestamps2 = [];
      for (const day of days) {
        const dayBase = monthStartMs + (day - 1) * MS_PER_DAY;
        for (const offset of timeSlotOffsets) {
          timestamps2.push(dayBase + offset);
        }
      }
      return this.applyBySetPosToSortedList(timestamps2).sort((a, b) => a - b);
    }
    const timestamps = [];
    for (const day of days) {
      const dayBase = monthStartMs + (day - 1) * MS_PER_DAY;
      for (const offset of timeSlotOffsets) {
        timestamps.push(dayBase + offset);
      }
    }
    return timestamps;
  }
  _allUtcMonthlyByDayOrMonthDay() {
    var _a;
    const dates = [];
    const startMilliseconds = this.originalDtstart.epochMilliseconds;
    const untilMilliseconds = (_a = this.opts.until) == null ? void 0 : _a.epochMilliseconds;
    let iterationCount = 0;
    if (!this.addDtstartIfNeeded(dates)) {
      return dates;
    }
    let monthIndex = this.originalDtstart.year * 12 + (this.originalDtstart.month - 1);
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      const { year, month } = this.monthIndexToYearMonth(monthIndex);
      const occurrenceEpochs = this.generateMonthlyOccurrenceEpochsUtc(year, month);
      for (const epochMilliseconds of occurrenceEpochs) {
        if (epochMilliseconds < startMilliseconds) {
          continue;
        }
        if (untilMilliseconds !== void 0 && epochMilliseconds > untilMilliseconds) {
          return dates;
        }
        dates.push(this.utcZdtFromEpochMilliseconds(epochMilliseconds));
        if (this.shouldBreakForCountLimit(dates.length)) {
          return dates;
        }
      }
      monthIndex += this.opts.interval;
    }
  }
  processOccurrences(occs, dates, start, iterator, extraFilters) {
    let shouldBreak = false;
    for (const occ of occs) {
      if (import_polyfill.Temporal.ZonedDateTime.compare(occ, start) < 0) continue;
      if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(occ, this.opts.until) > 0) {
        shouldBreak = true;
        break;
      }
      if (extraFilters && !extraFilters(occ)) {
        continue;
      }
      if (iterator && this.isExcluded(occ)) {
        continue;
      }
      if (iterator && !iterator(occ, dates.length)) {
        shouldBreak = true;
        break;
      }
      dates.push(occ);
      if (this.shouldBreakForCountLimit(dates.length)) {
        shouldBreak = true;
        break;
      }
    }
    return { shouldBreak };
  }
  /**
   * Returns all occurrences of the rule.
   * @param iterator - An optional callback iterator function that can be used to filter or modify the occurrences.
   * @returns An array of Temporal.ZonedDateTime objects representing all occurrences of the rule.
   */
  _allMonthlyByDayOrMonthDay(iterator) {
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    let monthCursor = start.with({ day: 1 });
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      let occs = this.generateMonthlyOccurrencesOptimizedBySetPos(monthCursor);
      if (!occs) {
        occs = this.generateMonthlyOccurrences(monthCursor);
        occs = this.applyBySetPos(occs);
      }
      const { shouldBreak } = this.processOccurrences(occs, dates, start, iterator);
      if (shouldBreak) {
        break;
      }
      monthCursor = monthCursor.add({ months: this.opts.interval });
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allWeekly(iterator) {
    var _a, _b, _c;
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    const dayMap = weekdayToIsoDay;
    const dows = this.opts.byDay ? [...(_a = this.allByDayIsoDays) != null ? _a : []] : this.opts.byMonthDay && this.opts.byMonthDay.length > 0 ? [...Object.values(dayMap)] : [start.dayOfWeek];
    const firstWeekDates = dows.map((dw) => {
      const delta = (dw - start.dayOfWeek + 7) % 7;
      return start.add({ days: delta });
    });
    const firstOccurrence = firstWeekDates.reduce((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b) <= 0 ? a : b);
    const wkstToken = (_b = extractWeekdayToken(this.opts.wkst || "MO")) != null ? _b : "MO";
    const wkstDay = (_c = dayMap[wkstToken]) != null ? _c : 1;
    const firstOccWeekOffset = (firstOccurrence.dayOfWeek - wkstDay + 7) % 7;
    let weekCursor = firstOccurrence.subtract({ days: firstOccWeekOffset });
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      let occs = dows.flatMap((dw) => {
        const delta = (dw - wkstDay + 7) % 7;
        const sameDate = weekCursor.add({ days: delta });
        return this.expandByTime(sameDate);
      }).sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
      occs = this.applyBySetPos(occs);
      const { shouldBreak } = this.processOccurrences(
        occs,
        dates,
        start,
        iterator,
        (occ) => this.matchesByMonth(occ) && this.matchesByMonthDay(occ)
      );
      if (shouldBreak) {
        break;
      }
      weekCursor = weekCursor.add({ weeks: this.opts.interval });
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allMonthlyByMonth(iterator) {
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    const months = this.opts.byMonth.filter((v) => typeof v === "number").sort((a, b) => a - b);
    let monthOffset = 0;
    let startMonthIndex = months.findIndex((m) => m >= start.month);
    if (startMonthIndex === -1) {
      startMonthIndex = 0;
      monthOffset = 1;
    }
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      const monthIndex = startMonthIndex + monthOffset;
      const targetMonth = months[monthIndex % months.length];
      const yearsToAdd = Math.floor(monthIndex / months.length);
      const candidate = start.with({
        year: start.year + yearsToAdd,
        month: targetMonth
      });
      if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(candidate, this.opts.until) > 0) {
        break;
      }
      if (import_polyfill.Temporal.ZonedDateTime.compare(candidate, start) >= 0) {
        if (iterator && this.isExcluded(candidate)) {
          continue;
        }
        if (iterator && !iterator(candidate, dates.length)) {
          break;
        }
        dates.push(candidate);
        if (this.shouldBreakForCountLimit(dates.length)) {
          break;
        }
      }
      monthOffset++;
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allYearlyByMonth(iterator) {
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    const months = this.opts.byMonth.filter((v) => typeof v === "number").sort((a, b) => a - b);
    let yearOffset = 0;
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      const year = start.year + yearOffset * this.opts.interval;
      for (const month of months) {
        let occ = start.with({ year, month });
        occ = this.applyTimeOverride(occ);
        if (import_polyfill.Temporal.ZonedDateTime.compare(occ, start) < 0) {
          continue;
        }
        if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(occ, this.opts.until) > 0) {
          return this.applyCountLimitAndMergeRDates(dates, iterator);
        }
        if (iterator && this.isExcluded(occ)) {
          continue;
        }
        if (iterator && !iterator(occ, dates.length)) {
          return this.applyCountLimitAndMergeRDates(dates, iterator);
        }
        dates.push(occ);
        if (this.shouldBreakForCountLimit(dates.length)) {
          return this.applyCountLimitAndMergeRDates(dates, iterator);
        }
      }
      yearOffset++;
    }
  }
  _allYearlyComplex(iterator) {
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    let yearCursor = start.with({ month: 1, day: 1 });
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      const occs = this.generateYearlyOccurrences(yearCursor);
      const uniqueOccs = [];
      if (occs.length > 0) {
        uniqueOccs.push(occs[0]);
        for (let i = 1; i < occs.length; i++) {
          if (import_polyfill.Temporal.ZonedDateTime.compare(occs[i], occs[i - 1]) !== 0) {
            uniqueOccs.push(occs[i]);
          }
        }
      }
      const { shouldBreak } = this.processOccurrences(uniqueOccs, dates, start, iterator);
      if (shouldBreak) {
        break;
      }
      const interval = this.opts.freq === "WEEKLY" ? 1 : this.opts.interval;
      yearCursor = yearCursor.add({ years: interval });
      if (this.opts.freq === "WEEKLY" && this.opts.until && yearCursor.year > this.opts.until.year) {
        break;
      }
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allMinutelySecondlyComplex(iterator) {
    const dates = [];
    let iterationCount = 0;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    let current = this.computeFirst();
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(current, this.opts.until) > 0) {
        break;
      }
      if (this.matchesAll(current)) {
        if (iterator && this.isExcluded(current)) {
          current = this.nextCandidateSameDate(current);
          continue;
        }
        if (iterator && !iterator(current, dates.length)) {
          break;
        }
        dates.push(current);
        if (this.shouldBreakForCountLimit(dates.length)) {
          break;
        }
        current = this.nextCandidateSameDate(current);
      } else {
        current = this.findNextValidDate(current);
      }
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allMonthlyByWeekNo(iterator) {
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    let current = start;
    const weekNos = [...this.opts.byWeekNo].sort((a, b) => a - b);
    const interval = this.opts.interval;
    let monthsAdvanced = 0;
    let lastYearProcessed = -1;
    outer_loop: while (true) {
      if (this.shouldBreakForCountLimit(dates.length)) {
        break;
      }
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      const year = current.year;
      if (year !== lastYearProcessed && current.month >= start.month) {
        lastYearProcessed = year;
        for (const weekNo of weekNos) {
          const occs = this.generateOccurrencesForWeekInYear(year, weekNo);
          for (const occ of occs) {
            if (import_polyfill.Temporal.ZonedDateTime.compare(occ, start) >= 0) {
              if (iterator && this.isExcluded(occ)) {
                continue;
              }
              if (iterator && !iterator(occ, dates.length)) {
                break outer_loop;
              }
              dates.push(occ);
              if (this.shouldBreakForCountLimit(dates.length)) {
                break outer_loop;
              }
            }
          }
        }
      }
      monthsAdvanced += interval;
      current = start.add({ months: monthsAdvanced });
      if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(current, this.opts.until) > 0) {
        break;
      }
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allMonthlyByYearDay(iterator) {
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    let year = start.year;
    const yearDays = [...this.opts.byYearDay].sort((a, b) => a - b);
    const interval = this.opts.interval;
    const startMonthAbs = start.year * 12 + start.month;
    outer_loop: while (true) {
      if (this.shouldBreakForCountLimit(dates.length)) {
        break;
      }
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      const yearStart = start.with({ year, month: 1, day: 1 });
      const lastDayOfYear = yearStart.with({ month: 12, day: 31 }).dayOfYear;
      for (const yd of yearDays) {
        const dayNum = yd > 0 ? yd : lastDayOfYear + yd + 1;
        if (dayNum <= 0 || dayNum > lastDayOfYear) continue;
        const baseOcc = yearStart.add({ days: dayNum - 1 });
        for (const occ of this.expandByTime(baseOcc)) {
          if (import_polyfill.Temporal.ZonedDateTime.compare(occ, start) < 0) continue;
          if (dates.some((d) => import_polyfill.Temporal.ZonedDateTime.compare(d, occ) === 0)) continue;
          const occMonthAbs = occ.year * 12 + occ.month;
          if ((occMonthAbs - startMonthAbs) % interval !== 0) {
            continue;
          }
          if (!this.matchesByMonth(occ)) {
            continue;
          }
          if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(occ, this.opts.until) > 0) {
            break outer_loop;
          }
          if (iterator && this.isExcluded(occ)) {
            continue;
          }
          if (iterator && !iterator(occ, dates.length)) {
            break outer_loop;
          }
          dates.push(occ);
          if (this.shouldBreakForCountLimit(dates.length)) {
            break outer_loop;
          }
        }
      }
      year++;
      if (this.opts.until && year > this.opts.until.year + 2) {
        break;
      }
      if (!this.opts.until && this.opts.count) {
        const yearsToScan = Math.ceil(this.opts.count / (this.opts.byYearDay.length || 1)) * interval + 5;
        if (year > start.year + yearsToScan) {
          break;
        }
      }
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allDailyMinutelyHourlyWithBySetPos(iterator) {
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    let cursor;
    let duration;
    switch (this.opts.freq) {
      case "MINUTELY":
        cursor = start.with({ second: 0, microsecond: 0, nanosecond: 0 });
        duration = { minutes: this.opts.interval };
        break;
      case "HOURLY":
        cursor = start.with({ minute: 0, second: 0, microsecond: 0, nanosecond: 0 });
        duration = { hours: this.opts.interval };
        break;
      case "DAILY":
        cursor = start.with({ hour: 0, minute: 0, second: 0, microsecond: 0, nanosecond: 0 });
        duration = { days: this.opts.interval };
        break;
      default:
        return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      let periodOccs = this.expandByTime(cursor);
      periodOccs = periodOccs.filter((occ) => this.matchesAll(occ));
      periodOccs = this.applyBySetPos(periodOccs);
      const { shouldBreak } = this.processOccurrences(periodOccs, dates, start, iterator);
      if (shouldBreak) {
        break;
      }
      cursor = cursor.add(duration);
      if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(cursor, this.opts.until) > 0) {
        break;
      }
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  _allFallback(iterator) {
    const dates = [];
    let iterationCount = 0;
    let current = this.computeFirst();
    if (this.includeDtstart && import_polyfill.Temporal.ZonedDateTime.compare(current, this.originalDtstart) > 0) {
      if (iterator && this.isExcluded(this.originalDtstart)) {
      } else {
        if (iterator && !iterator(this.originalDtstart, dates.length)) {
          return this.applyCountLimitAndMergeRDates(dates, iterator);
        }
        dates.push(this.originalDtstart);
        if (this.shouldBreakForCountLimit(dates.length)) {
          return this.applyCountLimitAndMergeRDates(dates, iterator);
        }
      }
    }
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(current, this.opts.until) > 0) {
        break;
      }
      if (this.matchesAll(current)) {
        if (iterator && this.isExcluded(current)) {
        } else {
          if (iterator && !iterator(current, dates.length)) {
            break;
          }
          dates.push(current);
          if (this.shouldBreakForCountLimit(dates.length)) {
            break;
          }
        }
      }
      current = this.nextCandidateSameDate(current);
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  /**
   * Returns all occurrences of the rule.
   * @param iterator - An optional callback iterator function that can be used to filter or modify the occurrences.
   * @returns An array of Temporal.ZonedDateTime objects representing all occurrences of the rule.
   */
  all(iterator) {
    if (this.opts.rscale && ["CHINESE", "HEBREW", "INDIAN"].includes(this.opts.rscale)) {
      if (["YEARLY", "MONTHLY", "WEEKLY"].includes(this.opts.freq) || !!this.opts.byYearDay || !!this.opts.byWeekNo || this.opts.byMonthDay && this.opts.byMonthDay.length > 0) {
        return this._allRscaleNonGregorian(iterator);
      }
    }
    if (this.opts.byWeekNo && this.opts.byYearDay) {
      const yearStart = this.originalDtstart.with({ month: 1, day: 1, hour: 0, minute: 0, second: 0, millisecond: 0 });
      const yearDays = this.opts.byYearDay.map((yd) => {
        const lastDayOfYear = yearStart.with({ month: 12, day: 31 }).dayOfYear;
        return yd > 0 ? yd : lastDayOfYear + yd + 1;
      });
      let possibleDate = false;
      for (const yd of yearDays) {
        const date = yearStart.add({ days: yd - 1 });
        if (this.matchesByWeekNo(date)) {
          possibleDate = true;
          break;
        }
      }
      if (!possibleDate) {
        return [];
      }
    }
    if (!this.opts.count && !this.opts.until && !iterator) {
      throw new Error("all() requires iterator when no COUNT/UNTIL");
    }
    const utcFastPathDates = this.allUtcFastPath(iterator);
    if (utcFastPathDates) {
      return utcFastPathDates;
    }
    if (this.opts.freq === "MONTHLY" && (this.opts.byDay || this.opts.byMonthDay) && !this.opts.byWeekNo) {
      return this._allMonthlyByDayOrMonthDay(iterator);
    }
    if (this.opts.freq === "WEEKLY" && !(this.opts.byYearDay && this.opts.byYearDay.length > 0) && !(this.opts.byWeekNo && this.opts.byWeekNo.length > 0)) {
      return this._allWeekly(iterator);
    }
    if (this.opts.freq === "MONTHLY" && this.opts.byMonth && !this.opts.byDay && !this.opts.byMonthDay && !this.opts.byYearDay) {
      return this._allMonthlyByMonth(iterator);
    }
    if (this.opts.freq === "YEARLY" && this.opts.byMonth && !this.opts.byDay && !this.opts.byMonthDay && !this.opts.byYearDay && !this.opts.byWeekNo) {
      return this._allYearlyByMonth(iterator);
    }
    if (this.opts.freq === "YEARLY" && (this.opts.byDay || this.opts.byMonthDay || this.opts.byYearDay || this.opts.byWeekNo) || this.opts.freq === "WEEKLY" && this.opts.byYearDay && this.opts.byYearDay.length > 0 || this.opts.freq === "WEEKLY" && this.opts.byWeekNo && this.opts.byWeekNo.length > 0) {
      return this._allYearlyComplex(iterator);
    }
    if ((this.opts.freq === "MINUTELY" || this.opts.freq === "SECONDLY") && (this.opts.byMonth || this.opts.byWeekNo || this.opts.byYearDay || this.opts.byMonthDay || this.opts.byDay)) {
      return this._allMinutelySecondlyComplex(iterator);
    }
    if (this.opts.freq === "MONTHLY" && this.opts.byWeekNo && this.opts.byWeekNo.length > 0) {
      return this._allMonthlyByWeekNo(iterator);
    }
    if (this.opts.freq === "MONTHLY" && this.opts.byYearDay && this.opts.byYearDay.length > 0 && !this.opts.byDay && !this.opts.byMonthDay) {
      return this._allMonthlyByYearDay(iterator);
    }
    if (this.opts.rscale && this.opts.freq === "MONTHLY" && !this.opts.byDay && !this.opts.byMonthDay && !this.opts.byWeekNo && !this.opts.byYearDay) {
      return this._allMonthlyRscaleSimple(iterator);
    }
    if ((this.opts.freq === "MINUTELY" || this.opts.freq === "HOURLY" || this.opts.freq === "DAILY") && this.opts.bySetPos) {
      return this._allDailyMinutelyHourlyWithBySetPos(iterator);
    }
    return this._allFallback(iterator);
  }
  /**
   * RFC 7529: RSCALE present, simple monthly iteration with SKIP behavior.
   * Handles month-to-month stepping from DTSTART's year/month aiming for DTSTART's day-of-month.
   * Applies SKIP=OMIT (skip invalid months), BACKWARD (clamp to last day), FORWARD (first day of next month).
   */
  _allMonthlyRscaleSimple(iterator) {
    var _a;
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    const interval = (_a = this.opts.interval) != null ? _a : 1;
    const targetDom = start.day;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    let cursor = start.with({ day: 1 });
    while (true) {
      if (++iterationCount > this.maxIterations) {
        throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
      }
      const lastDay = cursor.add({ months: 1 }).subtract({ days: 1 }).day;
      let occ = null;
      if (targetDom <= lastDay) {
        occ = cursor.with({ day: targetDom });
      } else {
        const skip = this.opts.skip || "OMIT";
        if (skip === "BACKWARD") {
          occ = cursor.with({ day: lastDay });
        } else if (skip === "FORWARD") {
          occ = cursor.add({ months: 1 }).with({ day: 1 });
        } else {
          occ = null;
        }
      }
      if (occ) {
        occ = occ.with({ hour: start.hour, minute: start.minute, second: start.second });
        if (!(iterator && this.isExcluded(occ))) {
          if (import_polyfill.Temporal.ZonedDateTime.compare(occ, start) >= 0) {
            if (!iterator || iterator(occ, dates.length)) {
              dates.push(occ);
              if (this.shouldBreakForCountLimit(dates.length)) break;
            } else {
              break;
            }
          }
        }
      }
      cursor = cursor.add({ months: interval });
      if (this.opts.until && import_polyfill.Temporal.ZonedDateTime.compare(cursor, this.opts.until) > 0) {
        break;
      }
    }
    return this.applyCountLimitAndMergeRDates(dates, iterator);
  }
  /**
   * Converts rDate entries to ZonedDateTime and merges with existing dates.
   * @param dates - Array of dates to merge with
   * @returns Merged and deduplicated array of dates
   */
  mergeAndDeduplicateRDates(dates) {
    if (this.opts.rDate) {
      dates.push(...this.opts.rDate);
    }
    dates.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
    const dedup = [];
    for (const d of dates) {
      if (dedup.length === 0 || import_polyfill.Temporal.ZonedDateTime.compare(d, dedup[dedup.length - 1]) !== 0) {
        dedup.push(d);
      }
    }
    return dedup;
  }
  /**
   * Checks if a date is in the exDate list.
   * @param date - Date to check
   * @returns True if the date is excluded
   */
  isExcluded(date) {
    if (!this.opts.exDate || this.opts.exDate.length === 0) return false;
    return this.opts.exDate.some((exDate) => import_polyfill.Temporal.ZonedDateTime.compare(date, exDate) === 0);
  }
  /**
   * Excludes exDate entries from the given array of dates.
   * @param dates - Array of dates to filter
   * @returns Filtered array with exDate entries removed
   */
  excludeExDates(dates) {
    if (!this.opts.exDate || this.opts.exDate.length === 0) return dates;
    return dates.filter((date) => {
      return !this.isExcluded(date);
    });
  }
  /**
   * Applies count limit and merges rDates with the rule-generated dates.
   * @param dates - Array of dates generated by the rule
   * @param iterator - Optional iterator function
   * @returns Final array of dates after merging and applying count limit
   */
  applyCountLimitAndMergeRDates(dates, iterator) {
    const merged = this.mergeAndDeduplicateRDates(dates);
    const excluded = this.excludeExDates(merged);
    const hasCountLimit = this.opts.count !== void 0;
    if (!hasCountLimit && !iterator) {
      return excluded;
    }
    let emitted = 0;
    const max = hasCountLimit ? this.opts.count : Infinity;
    const finalDates = [];
    for (const d of excluded) {
      if (emitted >= max) break;
      if (iterator && !iterator(d, emitted)) break;
      finalDates.push(d);
      emitted++;
    }
    return finalDates;
  }
  /**
   * Checks if the count limit should break the loop based on rDate presence.
   * @param matchCount - Current number of matches
   * @returns true if the loop should break
   */
  shouldBreakForCountLimit(matchCount) {
    if (this.opts.count === void 0) return false;
    if (!this.opts.rDate) {
      return matchCount >= this.opts.count;
    }
    const rDateCount = this.opts.rDate.length;
    const targetRuleCount = Math.max(this.opts.count - rDateCount, 0);
    const safetyMargin = Math.min(targetRuleCount, 10);
    return matchCount >= targetRuleCount + safetyMargin;
  }
  hasTimeOfDayBetween(startTime, endTime) {
    var _a, _b, _c;
    if (import_polyfill.Temporal.PlainTime.compare(startTime, endTime) >= 0) return false;
    const base = this.originalDtstart;
    const hours = (_a = this.opts.byHour) != null ? _a : [base.hour];
    const minutes = (_b = this.opts.byMinute) != null ? _b : [base.minute];
    const seconds = (_c = this.opts.bySecond) != null ? _c : [base.second];
    for (const hour of hours) {
      for (const minute of minutes) {
        for (const second of seconds) {
          const candidate = import_polyfill.Temporal.PlainTime.from({
            hour,
            minute,
            second,
            millisecond: base.millisecond,
            microsecond: base.microsecond,
            nanosecond: base.nanosecond
          });
          if (import_polyfill.Temporal.PlainTime.compare(candidate, startTime) >= 0 && import_polyfill.Temporal.PlainTime.compare(candidate, endTime) < 0) {
            return true;
          }
        }
      }
    }
    return false;
  }
  /**
   * Returns all occurrences of the rule within a specified time window.
   * @param after - The start date or Temporal.ZonedDateTime object.
   * @param before - The end date or Temporal.ZonedDateTime object.
   * @param inc - Optional boolean flag to include the end date in the results.
   * @returns An array of Temporal.ZonedDateTime objects representing all occurrences of the rule within the specified time window.
   */
  between(after, before, inc = false) {
    var _a;
    const startInst = after instanceof Date ? import_polyfill.Temporal.Instant.from(after.toISOString()) : after.toInstant();
    const endInst = before instanceof Date ? import_polyfill.Temporal.Instant.from(before.toISOString()) : before.toInstant();
    const startZdt = import_polyfill.Temporal.Instant.from(startInst).toZonedDateTimeISO(this.tzid);
    const beforeZdt = import_polyfill.Temporal.Instant.from(endInst).toZonedDateTimeISO(this.tzid);
    const tempOpts = __spreadValues({}, this.opts);
    if (!tempOpts.until || import_polyfill.Temporal.ZonedDateTime.compare(beforeZdt, tempOpts.until) < 0) {
      tempOpts.until = beforeZdt;
    }
    if (tempOpts.count === void 0) {
      const interval = (_a = tempOpts.interval) != null ? _a : 1;
      const aligned = startZdt.withPlainTime(this.originalDtstart.toPlainTime());
      let unit;
      switch (tempOpts.freq) {
        case "YEARLY":
          unit = "years";
          break;
        case "MONTHLY":
          unit = "months";
          break;
        case "WEEKLY":
          unit = "weeks";
          break;
        case "DAILY":
          unit = "days";
          break;
        case "HOURLY":
          unit = "hours";
          break;
        case "MINUTELY":
          unit = "minutes";
          break;
        default:
          unit = "seconds";
      }
      const dtstartNormalized = _RRuleTemporal.normalizeToPolyfill(this.opts.dtstart);
      const startZdtNormalized = _RRuleTemporal.normalizeToPolyfill(startZdt).withTimeZone(dtstartNormalized.timeZoneId);
      const alignedNormalized = _RRuleTemporal.normalizeToPolyfill(
        aligned.withPlainTime(this.originalDtstart.toPlainTime())
      ).withTimeZone(dtstartNormalized.timeZoneId);
      const diffAnchor = ["hours", "minutes", "seconds"].includes(unit) ? startZdtNormalized : alignedNormalized;
      const diffDur = dtstartNormalized.until(diffAnchor, { largestUnit: unit });
      const unitsBetween = diffDur[unit];
      let steps = Math.floor(unitsBetween / interval);
      const durationForJump = (jump) => {
        switch (unit) {
          case "years":
            return { years: jump };
          case "months":
            return { months: jump };
          case "weeks":
            return { weeks: jump };
          case "days":
            return { days: jump };
          case "hours":
            return { hours: jump };
          case "minutes":
            return { minutes: jump };
          default:
            return { seconds: jump };
        }
      };
      let candidate = _RRuleTemporal.normalizeToPolyfill(this.opts.dtstart.add(durationForJump(steps * interval)));
      if (steps > 0 && ["years", "months", "weeks", "days"].includes(unit)) {
        const sameDate = candidate.toPlainDate().equals(startZdtNormalized.toPlainDate());
        if (sameDate && import_polyfill.Temporal.ZonedDateTime.compare(candidate, startZdtNormalized) > 0) {
          if (this.hasTimeOfDayBetween(startZdtNormalized.toPlainTime(), candidate.toPlainTime())) {
            steps -= 1;
            candidate = _RRuleTemporal.normalizeToPolyfill(this.opts.dtstart.add(durationForJump(steps * interval)));
          }
        }
      }
      const dtstartForCompare = _RRuleTemporal.normalizeToPolyfill(this.opts.dtstart);
      if (import_polyfill.Temporal.ZonedDateTime.compare(candidate, dtstartForCompare) < 0) {
        candidate = dtstartForCompare;
      }
      tempOpts.dtstart = candidate;
    }
    const tempRule = new _RRuleTemporal(tempOpts);
    const allDates = tempRule.all();
    return allDates.filter((date) => {
      const inst = date.toInstant();
      const afterStart = inc ? import_polyfill.Temporal.Instant.compare(inst, startInst) >= 0 : import_polyfill.Temporal.Instant.compare(inst, startInst) > 0;
      const beforeEnd = inc ? import_polyfill.Temporal.Instant.compare(inst, endInst) <= 0 : import_polyfill.Temporal.Instant.compare(inst, endInst) < 0;
      return afterStart && beforeEnd;
    });
  }
  /**
   * Convenience helper: true if the exact instant is an occurrence of the rule.
   * This checks full date-time equality (including time and time zone).
   */
  matches(date) {
    return this.between(date, date, true).length > 0;
  }
  /**
   * Convenience helper: true if any occurrence falls on the given calendar day
   * in the rule's time zone. This ignores time-of-day granularity.
   */
  occursOn(date) {
    const startOfDay = date.toZonedDateTime({
      timeZone: this.tzid,
      plainTime: import_polyfill.Temporal.PlainTime.from("00:00")
    });
    const endOfDay = startOfDay.add({ days: 1 }).subtract({ nanoseconds: 1 });
    return this.between(startOfDay, endOfDay, true).length > 0;
  }
  /**
   * Returns the next occurrence of the rule after a specified date.
   * @param after - The start date or Temporal.ZonedDateTime object.
   * @param inc - Optional boolean flag to include occurrences on the start date.
   * @returns The next occurrence of the rule after the specified date or null if no occurrences are found.
   */
  next(after = /* @__PURE__ */ new Date(), inc = false) {
    const afterInst = after instanceof Date ? import_polyfill.Temporal.Instant.from(after.toISOString()) : after.toInstant();
    let result = null;
    this.all((occ) => {
      const inst = occ.toInstant();
      const ok = inc ? import_polyfill.Temporal.Instant.compare(inst, afterInst) >= 0 : import_polyfill.Temporal.Instant.compare(inst, afterInst) > 0;
      if (ok) {
        if (!result || import_polyfill.Temporal.ZonedDateTime.compare(occ, result) < 0) {
          result = occ;
        }
        return false;
      }
      return true;
    });
    return result;
  }
  /**
   * Returns the previous occurrence of the rule before a specified date.
   * @param before - The end date or Temporal.ZonedDateTime object.
   * @param inc - Optional boolean flag to include occurrences on the end date.
   * @returns The previous occurrence of the rule before the specified date or null if no occurrences are found.
   */
  previous(before = /* @__PURE__ */ new Date(), inc = false) {
    const beforeInst = before instanceof Date ? import_polyfill.Temporal.Instant.from(before.toISOString()) : before.toInstant();
    let prev = null;
    this.all((occ) => {
      const inst = occ.toInstant();
      const beyond = inc ? import_polyfill.Temporal.Instant.compare(inst, beforeInst) > 0 : import_polyfill.Temporal.Instant.compare(inst, beforeInst) >= 0;
      if (beyond) return false;
      prev = occ;
      return true;
    });
    return prev;
  }
  toString() {
    const iso = this.originalDtstart.toString({ smallestUnit: "second" }).replace(/[-:]/g, "");
    const dtLine = `DTSTART;TZID=${this.tzid}:${iso.slice(0, 15)}`;
    const rule = [];
    const {
      freq,
      interval,
      count,
      until,
      byHour,
      byMinute,
      bySecond,
      byDay,
      byMonth,
      byMonthDay,
      bySetPos,
      byWeekNo,
      byYearDay,
      wkst,
      rDate,
      exDate
    } = this.opts;
    if (this.opts.rscale) rule.push(`RSCALE=${this.opts.rscale}`);
    if (this.opts.rscale && this.opts.skip) rule.push(`SKIP=${this.opts.skip}`);
    rule.push(`FREQ=${freq}`);
    if (interval !== 1) rule.push(`INTERVAL=${interval}`);
    if (count !== void 0) rule.push(`COUNT=${count}`);
    if (until) {
      rule.push(`UNTIL=${this.formatIcsDateTime(until)}`);
    }
    if (byHour) rule.push(`BYHOUR=${byHour.join(",")}`);
    if (byMinute) rule.push(`BYMINUTE=${byMinute.join(",")}`);
    if (bySecond) rule.push(`BYSECOND=${bySecond.join(",")}`);
    if (byDay) rule.push(`BYDAY=${byDay.join(",")}`);
    if (byMonth) rule.push(`BYMONTH=${byMonth.join(",")}`);
    if (byMonthDay) rule.push(`BYMONTHDAY=${byMonthDay.join(",")}`);
    if (bySetPos) rule.push(`BYSETPOS=${bySetPos.join(",")}`);
    if (byWeekNo) rule.push(`BYWEEKNO=${byWeekNo.join(",")}`);
    if (byYearDay) rule.push(`BYYEARDAY=${byYearDay.join(",")}`);
    if (wkst) rule.push(`WKST=${wkst}`);
    const lines = [dtLine, `RRULE:${rule.join(";")}`];
    if (rDate) {
      lines.push(`RDATE:${this.joinDates(rDate)}`);
    }
    if (exDate) {
      lines.push(`EXDATE:${this.joinDates(exDate)}`);
    }
    return lines.join("\n");
  }
  formatIcsDateTime(date) {
    return date.toInstant().toString().replace(/[-:]/g, "").slice(0, 15) + "Z";
  }
  joinDates(dates) {
    return dates.map((d) => this.formatIcsDateTime(d));
  }
  /**
   * Given any date in a month, return all the ZonedDateTimes in that month
   * matching your opts.byDay and opts.byMonth (or the single "same day" if no BYDAY).
   */
  generateMonthlyOccurrences(sample) {
    const monthStart = sample.day === 1 ? sample : sample.with({ day: 1 });
    if (!this.opts.byDay && !this.opts.byMonthDay) {
      return this.expandByTime(sample);
    }
    const finalDays = this.generateMonthlyOccurrenceDays(monthStart);
    if (finalDays.length === 0) return [];
    const hits = finalDays.map((d) => monthStart.with({ day: d }));
    return hits.flatMap((z) => this.expandByTime(z));
  }
  /**
   * Given any date in a year, return all ZonedDateTimes in that year matching
   * the BYDAY/BYMONTHDAY/BYMONTH constraints. Months default to DTSTART's month
   * if BYMONTH is not specified.
   */
  generateYearlyOccurrences(sample) {
    const months = this.opts.byMonth ? this.opts.byMonth.filter((v) => typeof v === "number").sort((a, b) => a - b) : this.opts.byMonthDay || this.opts.byDay ? [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] : [this.originalDtstart.month];
    let occs = [];
    const hasOrdinalByDay = this.opts.byDay && this.opts.byDay.some((t) => /^[+-]?\d/.test(t));
    if (hasOrdinalByDay && !this.opts.byMonth) {
      const dayMap = weekdayToIsoDay;
      for (const tok of this.opts.byDay) {
        const parsed = parseByDayToken(tok);
        if (!parsed || parsed.ord === 0) continue;
        const ord = parsed.ord;
        const wd = dayMap[parsed.weekday];
        let dt;
        if (ord > 0) {
          const jan1 = sample.with({ month: 1, day: 1 });
          const delta = (wd - jan1.dayOfWeek + 7) % 7;
          dt = jan1.add({ days: delta + 7 * (ord - 1) });
        } else {
          const dec31 = sample.with({ month: 12, day: 31 });
          const delta = (dec31.dayOfWeek - wd + 7) % 7;
          dt = dec31.subtract({ days: delta + 7 * (-ord - 1) });
        }
        occs.push(...this.expandByTime(dt));
      }
    } else if (!this.opts.byYearDay && !this.opts.byWeekNo) {
      occs = [];
      for (const m of months) {
        const monthSample = sample.with({ month: m, day: 1 });
        const monthOccs = this.generateMonthlyOccurrences(monthSample);
        if (monthOccs.length === 0 && this.opts.rscale && this.opts.byMonthDay && this.opts.byMonthDay.length > 0) {
          const lastDay = monthSample.add({ months: 1 }).subtract({ days: 1 }).day;
          const target = this.opts.byMonthDay[0];
          const absTarget = target > 0 ? target : lastDay + target + 1;
          if (absTarget > lastDay || absTarget <= 0) {
            const skip = this.opts.skip || "OMIT";
            if (skip === "BACKWARD") {
              occs.push(...this.expandByTime(monthSample.with({ day: lastDay })));
            } else if (skip === "FORWARD") {
              const nextMonth = monthSample.add({ months: 1 }).with({ day: 1 });
              occs.push(...this.expandByTime(nextMonth));
            } else {
            }
          }
        } else {
          occs.push(...monthOccs);
        }
      }
    }
    if (this.opts.byYearDay) {
      const last = sample.with({ month: 12, day: 31 }).dayOfYear;
      for (const d of this.opts.byYearDay) {
        const dayNum = d > 0 ? d : last + d + 1;
        if (dayNum <= 0 || dayNum > last) continue;
        const dt = this.opts.freq === "MINUTELY" ? sample.with({ month: 1, day: 1, hour: 0, minute: 0, second: 0, millisecond: 0 }).add({ days: dayNum - 1 }) : sample.with({ month: 1, day: 1 }).add({ days: dayNum - 1 });
        if (!this.opts.byMonth || this.opts.byMonth.includes(dt.month)) {
          occs.push(...this.expandByTime(dt));
        }
      }
    }
    if (this.opts.byWeekNo) {
      const { lastWeek, firstWeekStart, tokens } = this.isoWeekByDay(sample);
      for (const weekNo of this.opts.byWeekNo) {
        if (weekNo > 0 && weekNo > lastWeek || weekNo < 0 && -weekNo > lastWeek) {
          continue;
        }
        const weekIndex = weekNo > 0 ? weekNo - 1 : lastWeek + weekNo;
        const weekStart = firstWeekStart.add({ weeks: weekIndex });
        occs.push(...this.addByDay(tokens, weekStart));
      }
    }
    occs = occs.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
    occs = this.applyBySetPos(occs);
    return occs;
  }
  addByDay(tokens, weekStart) {
    const dayMap = weekdayToIsoDay;
    const wkst = dayMap[this.opts.wkst || "MO"];
    const entries = [];
    for (const tok of tokens) {
      if (!tok) continue;
      const targetDow = dayMap[tok];
      const inst = weekStart.add({ days: (targetDow - wkst + 7) % 7 });
      if (!this.opts.byMonth || this.opts.byMonth.includes(inst.month)) {
        entries.push(...this.expandByTime(inst));
      }
    }
    return entries;
  }
  /**
   * Helper to find the next valid value from a sorted array
   */
  findNextValidValue(currentValue, validValues, compare) {
    return validValues.find((v) => compare(v, currentValue) > 0) || null;
  }
  /**
   * Efficiently find the next valid date for MINUTELY and SECONDLY frequency by jumping over
   * large gaps when BYXXX constraints don't match.
   */
  findNextValidDate(current) {
    if (this.opts.byWeekNo && this.opts.byYearDay) {
      const yearStart = current.with({ month: 1, day: 1, hour: 0, minute: 0, second: 0, millisecond: 0 });
      const yearDays = this.opts.byYearDay.map((yd) => {
        const lastDayOfYear = yearStart.with({ month: 12, day: 31 }).dayOfYear;
        return yd > 0 ? yd : lastDayOfYear + yd + 1;
      });
      for (const yd of yearDays) {
        const date = yearStart.add({ days: yd - 1 });
        if (this.matchesByWeekNo(date)) {
          break;
        }
      }
    }
    if (this.opts.byMonth) {
      const numericMonths = this.opts.byMonth.filter((v) => typeof v === "number");
      if (numericMonths.length && !numericMonths.includes(current.month)) {
        const months = [...numericMonths].sort((a, b) => a - b);
        const nextMonth = this.findNextValidValue(current.month, months, (a, b) => a - b);
        if (nextMonth) {
          current = current.with({ month: nextMonth, day: 1, hour: 0, minute: 0, second: 0 });
        } else {
          current = current.add({ years: 1 }).with({ month: months[0], day: 1, hour: 0, minute: 0, second: 0 });
        }
        current = this.applyTimeOverride(current);
        return current;
      }
    }
    if (this.opts.byWeekNo && !this.matchesByWeekNo(current)) {
      current = current.add({ weeks: 1 }).with({ hour: 0, minute: 0, second: 0 });
      current = this.applyTimeOverride(current);
      return current;
    }
    if (this.opts.byYearDay && !this.matchesByYearDay(current)) {
      const yearDays = [...this.opts.byYearDay].sort((a, b) => a - b);
      const currentYearDay = current.dayOfYear;
      const lastDayOfYear = current.with({ month: 12, day: 31 }).dayOfYear;
      let nextYearDay = yearDays.find((d) => {
        const dayNum = d > 0 ? d : lastDayOfYear + d + 1;
        return dayNum > currentYearDay;
      });
      if (nextYearDay) {
        const dayNum = nextYearDay > 0 ? nextYearDay : lastDayOfYear + nextYearDay + 1;
        if (this.opts.freq === "MINUTELY" || this.opts.freq === "SECONDLY") {
          current = current.with({ month: 1, day: 1, hour: 0, minute: 0, second: 0, millisecond: 0 }).add({ days: dayNum - 1 });
        } else {
          current = current.with({ month: 1, day: 1 }).add({ days: dayNum - 1 });
        }
      } else {
        const nextYear = current.add({ years: 1 });
        const nextYearLastDay = nextYear.with({ month: 12, day: 31 }).dayOfYear;
        const firstYearDay = yearDays[0];
        if (firstYearDay !== void 0) {
          const dayNum = firstYearDay > 0 ? firstYearDay : nextYearLastDay + firstYearDay + 1;
          if (this.opts.freq === "MINUTELY" || this.opts.freq === "SECONDLY") {
            current = nextYear.with({ month: 1, day: 1, hour: 0, minute: 0, second: 0, millisecond: 0 }).add({ days: dayNum - 1 });
          } else {
            current = nextYear.with({ month: 1, day: 1 }).add({ days: dayNum - 1 });
          }
        }
      }
      current = this.applyTimeOverride(current);
      return current;
    }
    if (this.opts.byMonthDay && !this.matchesByMonthDay(current)) {
      const monthDays = [...this.opts.byMonthDay].sort((a, b) => a - b);
      const lastDayOfMonth = current.with({ day: 1 }).add({ months: 1 }).subtract({ days: 1 }).day;
      const currentDay = current.day;
      const validDays = monthDays.map((d) => d > 0 ? d : lastDayOfMonth + d + 1).filter((d) => d > 0 && d <= lastDayOfMonth).sort((a, b) => a - b);
      const nextDay = this.findNextValidValue(currentDay, validDays, (a, b) => a - b);
      if (nextDay) {
        current = current.with({ day: nextDay, hour: 0, minute: 0, second: 0 });
      } else {
        const nextMonth = current.add({ months: 1 }).with({ day: 1 });
        const nextMonthLastDay = nextMonth.add({ months: 1 }).subtract({ days: 1 }).day;
        const firstMonthDay = monthDays[0];
        if (firstMonthDay !== void 0) {
          const dayNum = firstMonthDay > 0 ? firstMonthDay : nextMonthLastDay + firstMonthDay + 1;
          current = nextMonth.with({
            day: Math.max(1, Math.min(dayNum, nextMonthLastDay)),
            hour: 0,
            minute: 0,
            second: 0
          });
        } else {
          current = current.add({ months: 1 }).with({ day: 1, hour: 0, minute: 0, second: 0 });
        }
      }
      current = this.applyTimeOverride(current);
      return current;
    }
    if (this.opts.byDay && !this.matchesByDay(current)) {
      const targetDays = this.allByDayIsoDays;
      if (!(targetDays == null ? void 0 : targetDays.length)) {
        return this.applyTimeOverride(current.add({ days: 1 }).with({ hour: 0, minute: 0, second: 0 }));
      }
      const nextDayOfWeek = this.findNextValidValue(current.dayOfWeek, targetDays, (a, b) => a - b);
      if (nextDayOfWeek) {
        const delta = (nextDayOfWeek - current.dayOfWeek + 7) % 7;
        current = current.add({ days: delta }).with({ hour: 0, minute: 0, second: 0 });
      } else {
        const delta = (targetDays[0] - current.dayOfWeek + 7) % 7;
        current = current.add({ days: delta + 7 }).with({ hour: 0, minute: 0, second: 0 });
      }
      current = this.applyTimeOverride(current);
      return current;
    }
    switch (this.opts.freq) {
      case "SECONDLY":
      case "MINUTELY":
        current = current.add({ days: 1 }).with({ hour: 0, minute: 0, second: 0 });
        break;
      case "HOURLY":
        current = current.add({ days: 1 }).with({ hour: 0, minute: 0, second: 0 });
        break;
      case "DAILY":
      case "WEEKLY":
        current = current.add({ months: 1 }).with({ day: 1, hour: 0, minute: 0, second: 0 });
        break;
      case "MONTHLY":
      case "YEARLY":
        current = current.add({ years: 1 }).with({ month: 1, day: 1, hour: 0, minute: 0, second: 0 });
        break;
    }
    return this.applyTimeOverride(current);
  }
  applyBySetPos(list2) {
    const { bySetPos } = this.opts;
    if (!bySetPos || !bySetPos.length) return list2;
    const sorted = [...list2].sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
    const out = this.applyBySetPosToSortedList(sorted);
    return out.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
  }
  isoWeekByDay(sample) {
    var _a;
    const dayMap = weekdayToIsoDay;
    const wkst = dayMap[this.opts.wkst || "MO"];
    const jan1 = sample.with({ month: 1, day: 1 });
    const jan4 = sample.with({ month: 1, day: 4 });
    const delta = (jan4.dayOfWeek - wkst + 7) % 7;
    const firstWeekStart = jan4.subtract({ days: delta });
    const isLeapYear = jan1.inLeapYear;
    const lastWeek = jan1.dayOfWeek === 4 || isLeapYear && jan1.dayOfWeek === 3 ? 53 : 52;
    const tokens = ((_a = this.opts.byDay) == null ? void 0 : _a.length) ? this.opts.byDay.map((tok) => extractWeekdayToken(tok)).filter((day) => day !== null) : [Object.entries(dayMap).find(([, d]) => d === this.originalDtstart.dayOfWeek)[0]];
    return { lastWeek, firstWeekStart, tokens };
  }
  /**
   * Generate occurrences for a specific week number in a given year
   */
  generateOccurrencesForWeekInYear(year, weekNo) {
    const occs = [];
    const sample = this.originalDtstart.with({ year, month: 1, day: 1 });
    const { lastWeek, firstWeekStart, tokens } = this.isoWeekByDay(sample);
    if (weekNo > 0 && weekNo > lastWeek || weekNo < 0 && -weekNo > lastWeek) {
      return occs;
    }
    const weekIndex = weekNo > 0 ? weekNo - 1 : lastWeek + weekNo;
    const weekStart = firstWeekStart.add({ weeks: weekIndex });
    occs.push(...this.addByDay(tokens, weekStart));
    return occs.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
  }
  // ===== RSCALE (non-Gregorian) support: Chinese and Hebrew =====
  getRscaleCalendarId() {
    var _a;
    const map = {
      GREGORIAN: "gregory",
      CHINESE: "chinese",
      HEBREW: "hebrew",
      INDIAN: "indian"
    };
    const r = ((_a = this.opts.rscale) == null ? void 0 : _a.toUpperCase()) || "";
    return map[r] || null;
  }
  assertRscaleCalendarSupported(calId) {
    if (calId === "gregory" || calId === "iso8601") return;
    const cached = _RRuleTemporal.rscaleCalendarSupport[calId];
    if (cached === true) return;
    if (cached === false) {
      throw new Error(`RSCALE=${this.opts.rscale} is not supported by the current Temporal/Intl implementation`);
    }
    let supported = true;
    try {
      const probe = import_polyfill.Temporal.ZonedDateTime.from("2000-01-01T00:00:00+00:00[UTC]").withCalendar(calId);
      void probe.year;
      void probe.monthCode;
      void probe.day;
    } catch (e) {
      supported = false;
    }
    _RRuleTemporal.rscaleCalendarSupport[calId] = supported;
    if (!supported) {
      throw new Error(`RSCALE=${this.opts.rscale} is not supported by the current Temporal/Intl implementation`);
    }
  }
  pad2(n) {
    return String(n).padStart(2, "0");
  }
  monthMatchesToken(monthCode, token) {
    if (typeof token === "number") {
      return monthCode === `M${this.pad2(token)}`;
    }
    if (/^\d+L$/i.test(token)) {
      const n = parseInt(token, 10);
      return monthCode === `M${this.pad2(n)}L`;
    }
    return false;
  }
  monthsOfYear(calId, year) {
    const out = [];
    for (let m = 1; m <= 20; m++) {
      try {
        const d = import_polyfill.Temporal.PlainDate.from({ calendar: calId, year, month: m, day: 1 });
        out.push(d);
      } catch (e) {
        break;
      }
    }
    return out;
  }
  startOfYear(calId, year) {
    return import_polyfill.Temporal.PlainDate.from({ calendar: calId, year, month: 1, day: 1 });
  }
  endOfYear(calId, year) {
    return this.startOfYear(calId, year + 1).subtract({ days: 1 });
  }
  rscaleFirstWeekStart(calId, year, wkst) {
    const jan4 = import_polyfill.Temporal.PlainDate.from({ calendar: calId, year, month: 1, day: 4 });
    const delta = (jan4.dayOfWeek - wkst + 7) % 7;
    return jan4.subtract({ days: delta });
  }
  rscaleLastWeekCount(calId, year, wkst) {
    const firstWeekStart = this.rscaleFirstWeekStart(calId, year, wkst);
    const lastDay = this.endOfYear(calId, year);
    const diffDays = lastDay.since(firstWeekStart).days;
    return Math.floor(diffDays / 7) + 1;
  }
  lastDayOfMonth(pd) {
    return pd.with({ day: 1 }).add({ months: 1 }).subtract({ days: 1 }).day;
  }
  buildZdtFromPlainDate(pd) {
    const t = this.originalDtstart;
    const pdt = import_polyfill.Temporal.PlainDateTime.from({
      calendar: pd.calendarId,
      year: pd.year,
      month: pd.month,
      day: pd.day,
      hour: t.hour,
      minute: t.minute,
      second: t.second
    });
    return pdt.toZonedDateTime(this.tzid);
  }
  rscaleMatchesByYearDay(calId, pd) {
    const list2 = this.opts.byYearDay;
    if (!list2 || list2.length === 0) return true;
    const last = this.endOfYear(calId, pd.year).dayOfYear;
    return list2.some((d) => d > 0 ? pd.dayOfYear === d : pd.dayOfYear === last + d + 1);
  }
  rscaleMatchesByWeekNo(calId, pd) {
    const list2 = this.opts.byWeekNo;
    if (!list2 || list2.length === 0) return true;
    const dayMap = weekdayToIsoDay;
    const wkst = dayMap[this.opts.wkst || "MO"];
    const weekStart = pd.subtract({ days: (pd.dayOfWeek - wkst + 7) % 7 });
    const thursday = weekStart.add({ days: (4 - wkst + 7) % 7 });
    const weekYear = thursday.year;
    const firstStart = this.rscaleFirstWeekStart(calId, weekYear, wkst);
    const lastWeek = this.rscaleLastWeekCount(calId, weekYear, wkst);
    const idx = Math.floor(pd.since(firstStart).days / 7) + 1;
    return list2.some((wn) => wn > 0 ? idx === wn : idx === lastWeek + wn + 1);
  }
  rscaleMatchesByMonth(calId, pd) {
    const tokens = this.opts.byMonth;
    if (!tokens || tokens.length === 0) return true;
    return tokens.some((tok) => this.monthMatchesToken(pd.monthCode, tok));
  }
  rscaleMatchesByMonthDay(pd) {
    const list2 = this.opts.byMonthDay;
    if (!list2 || list2.length === 0) return true;
    const last = pd.with({ day: 1 }).add({ months: 1 }).subtract({ days: 1 }).day;
    const value = pd.day;
    return list2.some((d) => d > 0 ? value === d : value === last + d + 1);
  }
  rscaleMatchesByDayBasic(pd) {
    const byDay = this.opts.byDay;
    if (!byDay || byDay.length === 0) return true;
    const dayMap = weekdayToIsoDay;
    const tokens = byDay.map((tok) => extractWeekdayToken(tok)).filter((x) => x !== null);
    if (tokens.length === 0) return true;
    return tokens.some((wd) => dayMap[wd] === pd.dayOfWeek);
  }
  rscaleDateMatches(calId, pd) {
    return this.rscaleMatchesByMonth(calId, pd) && this.rscaleMatchesByYearDay(calId, pd) && this.rscaleMatchesByWeekNo(calId, pd) && this.rscaleMatchesByMonthDay(pd) && this.rscaleMatchesByDayBasic(pd);
  }
  applySkipForDay(calId, year, monthStart, targetDay) {
    const last = this.lastDayOfMonth(monthStart);
    const skip = this.opts.skip || "OMIT";
    if (targetDay >= 1 && targetDay <= last) {
      return monthStart.with({ day: targetDay });
    }
    if (skip === "BACKWARD") {
      return monthStart.with({ day: last });
    }
    if (skip === "FORWARD") {
      const nextMonthStart = monthStart.add({ months: 1 });
      return nextMonthStart.with({ day: 1 });
    }
    return null;
  }
  generateMonthlyOccurrencesRscale(calId, year, monthStart) {
    const occs = [];
    const byMonthDay = this.opts.byMonthDay;
    const byDay = this.opts.byDay;
    if (!byDay && !byMonthDay) {
      const targetDay = this.originalDtstart.withCalendar(calId).day;
      const pd = this.applySkipForDay(calId, year, monthStart, targetDay);
      if (pd) occs.push(this.buildZdtFromPlainDate(pd));
      return occs;
    }
    const addZ = (pd) => {
      occs.push(this.buildZdtFromPlainDate(pd));
    };
    const last = this.lastDayOfMonth(monthStart);
    const resolveDay = (d) => d > 0 ? d : last + d + 1;
    if (byMonthDay && byMonthDay.length > 0) {
      for (const raw of byMonthDay) {
        const dayNum = resolveDay(raw);
        const pd = this.applySkipForDay(calId, year, monthStart, dayNum);
        if (pd) addZ(pd);
      }
    }
    if (byDay && byDay.length > 0) {
      const dayMap = weekdayToIsoDay;
      const buckets = {};
      let cur = monthStart;
      while (cur.month === monthStart.month && cur.year === monthStart.year) {
        const wd = cur.dayOfWeek;
        (buckets[wd] || (buckets[wd] = [])).push(cur);
        cur = cur.add({ days: 1 });
      }
      for (const tok of byDay) {
        const parsed = parseByDayToken(tok);
        if (!parsed) continue;
        const ord = parsed.ord;
        const wd = dayMap[parsed.weekday];
        const list2 = buckets[wd] || [];
        if (list2.length === 0) continue;
        if (ord === 0) {
          for (const pd of list2) addZ(pd);
        } else {
          const idx = ord > 0 ? ord - 1 : list2.length + ord;
          const pd = list2[idx];
          if (pd) addZ(pd);
        }
      }
    }
    return this.applyBySetPos(occs).sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
  }
  _allRscaleNonGregorian(iterator) {
    var _a, _b, _c, _d, _e, _f, _g;
    const calId = this.getRscaleCalendarId();
    if (!calId) return this._allFallback(iterator);
    this.assertRscaleCalendarSupported(calId);
    const dates = [];
    let iterationCount = 0;
    const start = this.originalDtstart;
    const seed = start.withCalendar(calId);
    const interval = (_a = this.opts.interval) != null ? _a : 1;
    if (!this.addDtstartIfNeeded(dates, iterator)) {
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    if (this.opts.freq === "YEARLY") {
      let yearOffset = 0;
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        const tgtYear = seed.year + yearOffset * interval;
        let occs = [];
        const monthsTokens = this.opts.byMonth;
        const months = this.monthsOfYear(calId, tgtYear);
        const dayMap = weekdayToIsoDay;
        const wkst = dayMap[this.opts.wkst || "MO"];
        if (this.opts.byWeekNo && this.opts.byWeekNo.length > 0) {
          const firstStart = this.rscaleFirstWeekStart(calId, tgtYear, wkst);
          const lastWeek = this.rscaleLastWeekCount(calId, tgtYear, wkst);
          const tokens = ((_b = this.opts.byDay) == null ? void 0 : _b.length) ? this.opts.byDay.map((tok) => extractWeekdayToken(tok)).filter((day) => day !== null) : [Object.entries(dayMap).find(([, d]) => d === this.originalDtstart.dayOfWeek)[0]];
          for (const wn of this.opts.byWeekNo) {
            let idx = wn > 0 ? wn - 1 : lastWeek + wn;
            if (idx < 0 || idx >= lastWeek) continue;
            const weekStart = firstStart.add({ weeks: idx });
            for (const tok of tokens) {
              const targetDow = dayMap[tok];
              const pd = weekStart.add({ days: (targetDow - wkst + 7) % 7 });
              if (monthsTokens && monthsTokens.length > 0) {
                if (!monthsTokens.some((t) => this.monthMatchesToken(pd.monthCode, t))) continue;
              }
              if (this.opts.byYearDay && this.opts.byYearDay.length > 0) {
                const lastDay = this.endOfYear(calId, tgtYear).dayOfYear;
                const matches = this.opts.byYearDay.some((d) => {
                  const target = d > 0 ? d : lastDay + d + 1;
                  return pd.dayOfYear === target;
                });
                if (!matches) continue;
              }
              occs.push(this.buildZdtFromPlainDate(pd));
            }
          }
        } else if (this.opts.byYearDay && this.opts.byYearDay.length > 0) {
          const startOfYear = this.startOfYear(calId, tgtYear);
          const lastDay = this.endOfYear(calId, tgtYear).dayOfYear;
          for (const d of this.opts.byYearDay) {
            const target = d > 0 ? d : lastDay + d + 1;
            if (target < 1 || target > lastDay) continue;
            let pd = startOfYear.add({ days: target - 1 });
            if (monthsTokens && monthsTokens.length > 0) {
              if (!monthsTokens.some((t) => this.monthMatchesToken(pd.monthCode, t))) continue;
            }
            occs.push(this.buildZdtFromPlainDate(pd));
          }
        } else if (!monthsTokens || monthsTokens.length === 0) {
          try {
            const pd = import_polyfill.Temporal.PlainDate.from({
              calendar: calId,
              year: tgtYear,
              monthCode: seed.monthCode,
              day: seed.day
            });
            occs.push(this.buildZdtFromPlainDate(pd));
          } catch (e) {
            const skip = this.opts.skip || "OMIT";
            if (skip === "FORWARD" || skip === "BACKWARD") {
              const mapped = seed.with({ year: tgtYear });
              const adjusted = skip === "BACKWARD" ? mapped.subtract({ days: 1 }) : mapped;
              occs.push(adjusted.withCalendar("iso8601"));
            }
          }
        } else {
          const monthStarts = months.filter((m) => monthsTokens.some((tok) => this.monthMatchesToken(m.monthCode, tok)));
          for (const ms of monthStarts) {
            occs.push(...this.generateMonthlyOccurrencesRscale(calId, tgtYear, ms));
          }
        }
        if (occs.length > 0) {
          const expanded = occs.flatMap((z) => this.expandByTime(z));
          const sorted = expanded.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
          const { shouldBreak } = this.processOccurrences(sorted, dates, start, iterator);
          if (shouldBreak) break;
        }
        yearOffset++;
        if (this.opts.until && tgtYear > this.opts.until.withCalendar(calId).year) break;
      }
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    if (this.opts.freq === "WEEKLY") {
      const dayMap = weekdayToIsoDay;
      const wkst = dayMap[this.opts.wkst || "MO"];
      const tokens = ((_c = this.opts.byDay) == null ? void 0 : _c.length) ? this.opts.byDay.map((tok) => extractWeekdayToken(tok)).filter((day) => day !== null) : [Object.entries(dayMap).find(([, d]) => d === this.originalDtstart.dayOfWeek)[0]];
      let weekStart = seed.toPlainDate().subtract({ days: (seed.dayOfWeek - wkst + 7) % 7 });
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        const occs = [];
        for (const tok of tokens) {
          const targetDow = dayMap[tok];
          const pd = weekStart.add({ days: (targetDow - wkst + 7) % 7 });
          if (import_polyfill.Temporal.ZonedDateTime.compare(this.buildZdtFromPlainDate(pd), this.originalDtstart) < 0) {
            continue;
          }
          if (this.opts.byWeekNo && this.opts.byWeekNo.length > 0) {
            const thursday = weekStart.add({ days: (4 - wkst + 7) % 7 });
            const weekYear = thursday.year;
            const firstStart = this.rscaleFirstWeekStart(calId, weekYear, wkst);
            const lastWeek = this.rscaleLastWeekCount(calId, weekYear, wkst);
            const idx = Math.floor(pd.since(firstStart).days / 7) + 1;
            const match = this.opts.byWeekNo.some((wn) => wn > 0 ? idx === wn : idx === lastWeek + wn + 1);
            if (!match) continue;
          }
          if (this.opts.byYearDay && this.opts.byYearDay.length > 0) {
            const last = this.endOfYear(calId, pd.year).dayOfYear;
            const match = this.opts.byYearDay.some((d) => d > 0 ? pd.dayOfYear === d : pd.dayOfYear === last + d + 1);
            if (!match) continue;
          }
          const monthsTokens = this.opts.byMonth;
          if (monthsTokens && monthsTokens.length > 0) {
            if (!monthsTokens.some((t) => this.monthMatchesToken(pd.monthCode, t))) continue;
          }
          occs.push(this.buildZdtFromPlainDate(pd));
        }
        if (occs.length) {
          const expanded = occs.flatMap((z) => this.expandByTime(z));
          const sorted = expanded.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
          const { shouldBreak } = this.processOccurrences(sorted, dates, start, iterator);
          if (shouldBreak) return this.applyCountLimitAndMergeRDates(dates, iterator);
        }
        weekStart = weekStart.add({ weeks: (_d = this.opts.interval) != null ? _d : 1 });
        if (this.opts.until) {
          const z = this.buildZdtFromPlainDate(weekStart.add({ days: 6 }));
          if (import_polyfill.Temporal.ZonedDateTime.compare(z, this.opts.until) > 0) break;
        }
      }
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    if (this.opts.freq === "MONTHLY") {
      let cursor = seed.toPlainDate().with({ day: 1 });
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        const year = cursor.year;
        const monthStart = cursor;
        let proceed = true;
        const monthsTokens = this.opts.byMonth;
        if (monthsTokens && monthsTokens.length > 0) {
          proceed = monthsTokens.some((tok) => this.monthMatchesToken(monthStart.monthCode, tok));
        }
        if (proceed) {
          const occs = this.generateMonthlyOccurrencesRscale(calId, year, monthStart);
          const expanded = occs.flatMap((z) => this.expandByTime(z));
          const sorted = expanded.sort((a, b) => import_polyfill.Temporal.ZonedDateTime.compare(a, b));
          const { shouldBreak } = this.processOccurrences(sorted, dates, start, iterator);
          if (shouldBreak) break;
        }
        cursor = cursor.add({ months: (_e = this.opts.interval) != null ? _e : 1 });
        if (this.opts.until) {
          const z = this.buildZdtFromPlainDate(cursor);
          if (import_polyfill.Temporal.ZonedDateTime.compare(z, this.opts.until) > 0) break;
        }
      }
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    if (this.opts.freq === "DAILY") {
      let pd = seed.toPlainDate();
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        if (this.rscaleDateMatches(calId, pd)) {
          const base = this.buildZdtFromPlainDate(pd);
          let occs = this.expandByTime(base);
          occs = this.applyBySetPos(occs);
          const { shouldBreak } = this.processOccurrences(occs, dates, start, iterator);
          if (shouldBreak) break;
        }
        pd = pd.add({ days: (_f = this.opts.interval) != null ? _f : 1 });
        if (this.opts.until) {
          const z = this.buildZdtFromPlainDate(pd);
          if (import_polyfill.Temporal.ZonedDateTime.compare(z, this.opts.until) > 0) break;
        }
      }
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    if (this.opts.freq === "HOURLY" || this.opts.freq === "MINUTELY") {
      const unit = this.opts.freq === "HOURLY" ? "hour" : "minute";
      const unitMs = this.opts.freq === "HOURLY" ? 36e5 : 6e4;
      const interval2 = (_g = this.opts.interval) != null ? _g : 1;
      let pd = seed.toPlainDate();
      const startInstantMs = this.originalDtstart.toInstant().epochMilliseconds;
      while (true) {
        if (++iterationCount > this.maxIterations) {
          throw new Error(`Maximum iterations (${this.maxIterations}) exceeded in all()`);
        }
        if (this.rscaleDateMatches(calId, pd)) {
          const base = this.buildZdtFromPlainDate(pd);
          let occs = this.expandByTime(base);
          occs = occs.filter((occ) => {
            const delta = occ.toInstant().epochMilliseconds - startInstantMs;
            const steps = Math.floor(delta / unitMs);
            return steps % interval2 === 0;
          });
          const { shouldBreak } = this.processOccurrences(occs, dates, start, iterator);
          if (shouldBreak) break;
        }
        pd = pd.add({ days: 1 });
        if (this.opts.until) {
          const z = this.buildZdtFromPlainDate(pd);
          if (import_polyfill.Temporal.ZonedDateTime.compare(z, this.opts.until) > 0) break;
        }
      }
      return this.applyCountLimitAndMergeRDates(dates, iterator);
    }
    return this._allFallback(iterator);
  }
};
_RRuleTemporal.rscaleCalendarSupport = {};
var RRuleTemporal = _RRuleTemporal;

// src/totext.ts
var en = {
  weekdayNames: ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"],
  monthNames: [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December"
  ],
  units: {
    year: { singular: "year", plural: "years" },
    month: { singular: "month", plural: "months" },
    week: { singular: "week", plural: "weeks" },
    day: { singular: "day", plural: "days" },
    hour: { singular: "hour", plural: "hours" },
    minute: { singular: "minute", plural: "minutes" },
    second: { singular: "second", plural: "seconds" }
  },
  words: {
    every: "every",
    weekday: "weekday",
    on: "on",
    in: "in",
    on_the: "on the",
    day_of_month: "day of the month",
    day_of_year: "day of the year",
    in_week: "in week",
    at: "at",
    at_minute: "at minute",
    at_second: "at second",
    until: "until",
    for: "for",
    time: "time",
    times: "times",
    instance: "instance",
    week_starts_on: "week starts on",
    with: "with",
    additional_date: "additional date",
    additional_dates: "additional dates",
    excluding: "excluding",
    date: "date",
    dates: "dates",
    and: "and",
    last: "last",
    starting_from: "starting from"
  },
  ordinal: (n) => {
    const abs = Math.abs(n);
    const suffix = abs % 10 === 1 && abs % 100 !== 11 ? "st" : abs % 10 === 2 && abs % 100 !== 12 ? "nd" : abs % 10 === 3 && abs % 100 !== 13 ? "rd" : "th";
    return n < 0 ? `last` : `${abs}${suffix}`;
  }
};
var de = {
  weekdayNames: ["Montag", "Dienstag", "Mittwoch", "Donnerstag", "Freitag", "Samstag", "Sonntag"],
  monthNames: [
    "Januar",
    "Februar",
    "M\xE4rz",
    "April",
    "Mai",
    "Juni",
    "Juli",
    "August",
    "September",
    "Oktober",
    "November",
    "Dezember"
  ],
  units: {
    year: { singular: "Jahr", plural: "Jahre" },
    month: { singular: "Monat", plural: "Monate" },
    week: { singular: "Woche", plural: "Wochen" },
    day: { singular: "Tag", plural: "Tage" },
    hour: { singular: "Stunde", plural: "Stunden" },
    minute: { singular: "Minute", plural: "Minuten" },
    second: { singular: "Sekunde", plural: "Sekunden" }
  },
  words: {
    every: "jede/n/s",
    weekday: "Werktag",
    on: "am",
    in: "im",
    on_the: "am",
    day_of_month: "Tag des Monats",
    day_of_year: "Tag des Jahres",
    in_week: "in Kalenderwoche",
    at: "um",
    at_minute: "in Minute",
    at_second: "in Sekunde",
    until: "bis",
    for: "f\xFCr",
    time: "Mal",
    times: "Mal",
    instance: "Vorkommen",
    week_starts_on: "Woche beginnt am",
    with: "mit",
    additional_date: "zus\xE4tzlichem Datum",
    additional_dates: "zus\xE4tzlichen Daten",
    excluding: "ohne",
    date: "Datum",
    dates: "Daten",
    and: "und",
    last: "letzter",
    starting_from: "beginnend am"
  },
  ordinal: (n) => {
    if (n < 0) return "letzten";
    return `${Math.abs(n)}.`;
  }
};
var es = {
  weekdayNames: ["lunes", "martes", "mi\xE9rcoles", "jueves", "viernes", "s\xE1bado", "domingo"],
  monthNames: [
    "enero",
    "febrero",
    "marzo",
    "abril",
    "mayo",
    "junio",
    "julio",
    "agosto",
    "septiembre",
    "octubre",
    "noviembre",
    "diciembre"
  ],
  units: {
    year: { singular: "a\xF1o", plural: "a\xF1os" },
    month: { singular: "mes", plural: "meses" },
    week: { singular: "semana", plural: "semanas" },
    day: { singular: "d\xEDa", plural: "d\xEDas" },
    hour: { singular: "hora", plural: "horas" },
    minute: { singular: "minuto", plural: "minutos" },
    second: { singular: "segundo", plural: "segundos" }
  },
  words: {
    every: "cada",
    weekday: "d\xEDa de la semana",
    on: "en",
    in: "en",
    on_the: "el",
    day_of_month: "d\xEDa del mes",
    day_of_year: "d\xEDa del a\xF1o",
    in_week: "en la semana",
    at: "a las",
    at_minute: "en el minuto",
    at_second: "en el segundo",
    until: "hasta",
    for: "durante",
    time: "vez",
    times: "veces",
    instance: "ocasi\xF3n",
    week_starts_on: "la semana comienza el",
    with: "con",
    additional_date: "fecha adicional",
    additional_dates: "fechas adicionales",
    excluding: "excluyendo",
    date: "fecha",
    dates: "fechas",
    and: "y",
    last: "\xFAltimo",
    starting_from: "a partir de"
  },
  ordinal: (n) => n < 0 ? "\xFAltimo" : `${Math.abs(n)}\xBA`
};
var hi = {
  weekdayNames: ["\u0938\u094B\u092E\u0935\u093E\u0930", "\u092E\u0902\u0917\u0932\u0935\u093E\u0930", "\u092C\u0941\u0927\u0935\u093E\u0930", "\u0917\u0941\u0930\u0941\u0935\u093E\u0930", "\u0936\u0941\u0915\u094D\u0930\u0935\u093E\u0930", "\u0936\u0928\u093F\u0935\u093E\u0930", "\u0930\u0935\u093F\u0935\u093E\u0930"],
  monthNames: [
    "\u091C\u0928\u0935\u0930\u0940",
    "\u092B\u0930\u0935\u0930\u0940",
    "\u092E\u093E\u0930\u094D\u091A",
    "\u0905\u092A\u094D\u0930\u0948\u0932",
    "\u092E\u0908",
    "\u091C\u0942\u0928",
    "\u091C\u0941\u0932\u093E\u0908",
    "\u0905\u0917\u0938\u094D\u0924",
    "\u0938\u093F\u0924\u0902\u092C\u0930",
    "\u0905\u0915\u094D\u091F\u0942\u092C\u0930",
    "\u0928\u0935\u0902\u092C\u0930",
    "\u0926\u093F\u0938\u0902\u092C\u0930"
  ],
  units: {
    year: { singular: "\u0938\u093E\u0932", plural: "\u0938\u093E\u0932" },
    month: { singular: "\u092E\u0939\u0940\u0928\u093E", plural: "\u092E\u0939\u0940\u0928\u0947" },
    week: { singular: "\u0938\u092A\u094D\u0924\u093E\u0939", plural: "\u0938\u092A\u094D\u0924\u093E\u0939" },
    day: { singular: "\u0926\u093F\u0928", plural: "\u0926\u093F\u0928" },
    hour: { singular: "\u0918\u0902\u091F\u093E", plural: "\u0918\u0902\u091F\u0947" },
    minute: { singular: "\u092E\u093F\u0928\u091F", plural: "\u092E\u093F\u0928\u091F" },
    second: { singular: "\u0938\u0947\u0915\u0902\u0921", plural: "\u0938\u0947\u0915\u0902\u0921" }
  },
  words: {
    every: "\u0939\u0930",
    weekday: "\u0938\u092A\u094D\u0924\u093E\u0939 \u0915\u093E \u0926\u093F\u0928",
    on: "\u0915\u094B",
    in: "\u092E\u0947\u0902",
    on_the: "\u0915\u094B",
    day_of_month: "\u092E\u0939\u0940\u0928\u0947 \u0915\u093E \u0926\u093F\u0928",
    day_of_year: "\u0938\u093E\u0932 \u0915\u093E \u0926\u093F\u0928",
    in_week: "\u0938\u092A\u094D\u0924\u093E\u0939",
    at: "\u092A\u0930",
    at_minute: "\u092E\u093F\u0928\u091F \u092A\u0930",
    at_second: "\u0938\u0947\u0915\u0902\u0921 \u092A\u0930",
    until: "\u0924\u0915",
    for: "\u0915\u0947 \u0932\u093F\u090F",
    time: "\u092C\u093E\u0930",
    times: "\u092C\u093E\u0930",
    instance: "\u092C\u093E\u0930",
    week_starts_on: "\u0938\u092A\u094D\u0924\u093E\u0939 \u0936\u0941\u0930\u0942 \u0939\u094B\u0924\u093E \u0939\u0948",
    with: "\u0938\u093E\u0925",
    additional_date: "\u0905\u0924\u093F\u0930\u093F\u0915\u094D\u0924 \u0924\u093E\u0930\u0940\u0916",
    additional_dates: "\u0905\u0924\u093F\u0930\u093F\u0915\u094D\u0924 \u0924\u093E\u0930\u0940\u0916\u0947\u0902",
    excluding: "\u0915\u094B \u091B\u094B\u0921\u093C\u0915\u0930",
    date: "\u0924\u093E\u0930\u0940\u0916",
    dates: "\u0924\u093E\u0930\u0940\u0916\u0947\u0902",
    and: "\u0914\u0930",
    last: "\u0906\u0916\u093F\u0930\u0940",
    starting_from: "\u0938\u0947 \u0936\u0941\u0930\u0942"
  },
  ordinal: (n) => n < 0 ? "\u0906\u0916\u093F\u0930\u0940" : `${Math.abs(n)}\u0935\u093E\u0902`
};
var yue = {
  weekdayNames: ["\u661F\u671F\u4E00", "\u661F\u671F\u4E8C", "\u661F\u671F\u4E09", "\u661F\u671F\u56DB", "\u661F\u671F\u4E94", "\u661F\u671F\u516D", "\u661F\u671F\u65E5"],
  monthNames: ["\u4E00\u6708", "\u4E8C\u6708", "\u4E09\u6708", "\u56DB\u6708", "\u4E94\u6708", "\u516D\u6708", "\u4E03\u6708", "\u516B\u6708", "\u4E5D\u6708", "\u5341\u6708", "\u5341\u4E00\u6708", "\u5341\u4E8C\u6708"],
  units: {
    year: { singular: "\u5E74", plural: "\u5E74" },
    month: { singular: "\u6708", plural: "\u6708" },
    week: { singular: "\u9031", plural: "\u9031" },
    day: { singular: "\u65E5", plural: "\u65E5" },
    hour: { singular: "\u5C0F\u6642", plural: "\u5C0F\u6642" },
    minute: { singular: "\u5206\u9418", plural: "\u5206\u9418" },
    second: { singular: "\u79D2", plural: "\u79D2" }
  },
  words: {
    every: "\u6BCF",
    weekday: "\u5E73\u65E5",
    on: "\u5728",
    in: "\u65BC",
    on_the: "\u5728",
    day_of_month: "\u6708\u7684\u65E5\u5B50",
    day_of_year: "\u5E74\u7684\u65E5\u5B50",
    in_week: "\u7B2C",
    at: "\u5728",
    at_minute: "\u5728\u7B2C",
    at_second: "\u5728\u7B2C",
    until: "\u76F4\u5230",
    for: "\u5171",
    time: "\u6B21",
    times: "\u6B21",
    instance: "\u6B21",
    week_starts_on: "\u661F\u671F\u958B\u59CB\u65BC",
    with: "\u5E36\u6709",
    additional_date: "\u984D\u5916\u65E5\u671F",
    additional_dates: "\u984D\u5916\u65E5\u671F",
    excluding: "\u6392\u9664",
    date: "\u65E5\u671F",
    dates: "\u65E5\u671F",
    and: "\u548C",
    last: "\u6700\u5F8C",
    starting_from: "\u7531"
  },
  ordinal: (n) => n < 0 ? "\u6700\u5F8C" : `\u7B2C${Math.abs(n)}`
};
var ar = {
  weekdayNames: ["\u0627\u0644\u0627\u062B\u0646\u064A\u0646", "\u0627\u0644\u062B\u0644\u0627\u062B\u0627\u0621", "\u0627\u0644\u0623\u0631\u0628\u0639\u0627\u0621", "\u0627\u0644\u062E\u0645\u064A\u0633", "\u0627\u0644\u062C\u0645\u0639\u0629", "\u0627\u0644\u0633\u0628\u062A", "\u0627\u0644\u0623\u062D\u062F"],
  monthNames: [
    "\u064A\u0646\u0627\u064A\u0631",
    "\u0641\u0628\u0631\u0627\u064A\u0631",
    "\u0645\u0627\u0631\u0633",
    "\u0623\u0628\u0631\u064A\u0644",
    "\u0645\u0627\u064A\u0648",
    "\u064A\u0648\u0646\u064A\u0648",
    "\u064A\u0648\u0644\u064A\u0648",
    "\u0623\u063A\u0633\u0637\u0633",
    "\u0633\u0628\u062A\u0645\u0628\u0631",
    "\u0623\u0643\u062A\u0648\u0628\u0631",
    "\u0646\u0648\u0641\u0645\u0628\u0631",
    "\u062F\u064A\u0633\u0645\u0628\u0631"
  ],
  units: {
    year: { singular: "\u0633\u0646\u0629", plural: "\u0633\u0646\u0648\u0627\u062A" },
    month: { singular: "\u0634\u0647\u0631", plural: "\u0623\u0634\u0647\u0631" },
    week: { singular: "\u0623\u0633\u0628\u0648\u0639", plural: "\u0623\u0633\u0627\u0628\u064A\u0639" },
    day: { singular: "\u064A\u0648\u0645", plural: "\u0623\u064A\u0627\u0645" },
    hour: { singular: "\u0633\u0627\u0639\u0629", plural: "\u0633\u0627\u0639\u0627\u062A" },
    minute: { singular: "\u062F\u0642\u064A\u0642\u0629", plural: "\u062F\u0642\u0627\u0626\u0642" },
    second: { singular: "\u062B\u0627\u0646\u064A\u0629", plural: "\u062B\u0648\u0627\u0646\u064A" }
  },
  words: {
    every: "\u0643\u0644",
    weekday: "\u064A\u0648\u0645 \u0645\u0646 \u0623\u064A\u0627\u0645 \u0627\u0644\u0623\u0633\u0628\u0648\u0639",
    on: "\u0641\u064A",
    in: "\u0641\u064A",
    on_the: "\u0641\u064A \u0627\u0644\u0640",
    day_of_month: "\u064A\u0648\u0645 \u0645\u0646 \u0627\u0644\u0634\u0647\u0631",
    day_of_year: "\u064A\u0648\u0645 \u0645\u0646 \u0627\u0644\u0633\u0646\u0629",
    in_week: "\u0641\u064A \u0627\u0644\u0623\u0633\u0628\u0648\u0639",
    at: "\u0639\u0646\u062F",
    at_minute: "\u0641\u064A \u0627\u0644\u062F\u0642\u064A\u0642\u0629",
    at_second: "\u0641\u064A \u0627\u0644\u062B\u0627\u0646\u064A\u0629",
    until: "\u062D\u062A\u0649",
    for: "\u0644\u0645\u062F\u0629",
    time: "\u0645\u0631\u0629",
    times: "\u0645\u0631\u0627\u062A",
    instance: "\u0645\u0631\u0629",
    week_starts_on: "\u064A\u0628\u062F\u0623 \u0627\u0644\u0623\u0633\u0628\u0648\u0639 \u064A\u0648\u0645",
    with: "\u0645\u0639",
    additional_date: "\u062A\u0627\u0631\u064A\u062E \u0625\u0636\u0627\u0641\u064A",
    additional_dates: "\u062A\u0648\u0627\u0631\u064A\u062E \u0625\u0636\u0627\u0641\u064A\u0629",
    excluding: "\u0628\u0627\u0633\u062A\u062B\u0646\u0627\u0621",
    date: "\u062A\u0627\u0631\u064A\u062E",
    dates: "\u062A\u0648\u0627\u0631\u064A\u062E",
    and: "\u0648",
    last: "\u0627\u0644\u0623\u062E\u064A\u0631",
    starting_from: "\u0627\u0628\u062A\u062F\u0627\u0621\u064B \u0645\u0646"
  },
  ordinal: (n) => {
    if (n < 0) return "\u0627\u0644\u0623\u062E\u064A\u0631";
    const abs = Math.abs(n);
    const map = {
      1: "\u0627\u0644\u0623\u0648\u0644",
      2: "\u0627\u0644\u062B\u0627\u0646\u064A",
      3: "\u0627\u0644\u062B\u0627\u0644\u062B",
      4: "\u0627\u0644\u0631\u0627\u0628\u0639",
      5: "\u0627\u0644\u062E\u0627\u0645\u0633",
      6: "\u0627\u0644\u0633\u0627\u062F\u0633",
      7: "\u0627\u0644\u0633\u0627\u0628\u0639",
      8: "\u0627\u0644\u062B\u0627\u0645\u0646",
      9: "\u0627\u0644\u062A\u0627\u0633\u0639",
      10: "\u0627\u0644\u0639\u0627\u0634\u0631",
      11: "\u0627\u0644\u062D\u0627\u062F\u064A \u0639\u0634\u0631",
      12: "\u0627\u0644\u062B\u0627\u0646\u064A \u0639\u0634\u0631",
      13: "\u0627\u0644\u062B\u0627\u0644\u062B \u0639\u0634\u0631"
    };
    return map[abs] || abs.toString();
  }
};
var he = {
  weekdayNames: ["\u05D9\u05D5\u05DD \u05E9\u05E0\u05D9", "\u05D9\u05D5\u05DD \u05E9\u05DC\u05D9\u05E9\u05D9", "\u05D9\u05D5\u05DD \u05E8\u05D1\u05D9\u05E2\u05D9", "\u05D9\u05D5\u05DD \u05D7\u05DE\u05D9\u05E9\u05D9", "\u05D9\u05D5\u05DD \u05E9\u05D9\u05E9\u05D9", "\u05D9\u05D5\u05DD \u05E9\u05D1\u05EA", "\u05D9\u05D5\u05DD \u05E8\u05D0\u05E9\u05D5\u05DF"],
  monthNames: [
    "\u05D9\u05E0\u05D5\u05D0\u05E8",
    "\u05E4\u05D1\u05E8\u05D5\u05D0\u05E8",
    "\u05DE\u05E8\u05E5",
    "\u05D0\u05E4\u05E8\u05D9\u05DC",
    "\u05DE\u05D0\u05D9",
    "\u05D9\u05D5\u05E0\u05D9",
    "\u05D9\u05D5\u05DC\u05D9",
    "\u05D0\u05D5\u05D2\u05D5\u05E1\u05D8",
    "\u05E1\u05E4\u05D8\u05DE\u05D1\u05E8",
    "\u05D0\u05D5\u05E7\u05D8\u05D5\u05D1\u05E8",
    "\u05E0\u05D5\u05D1\u05DE\u05D1\u05E8",
    "\u05D3\u05E6\u05DE\u05D1\u05E8"
  ],
  units: {
    year: { singular: "\u05E9\u05E0\u05D4", plural: "\u05E9\u05E0\u05D9\u05DD" },
    month: { singular: "\u05D7\u05D5\u05D3\u05E9", plural: "\u05D7\u05D5\u05D3\u05E9\u05D9\u05DD" },
    week: { singular: "\u05E9\u05D1\u05D5\u05E2", plural: "\u05E9\u05D1\u05D5\u05E2\u05D5\u05EA" },
    day: { singular: "\u05D9\u05D5\u05DD", plural: "\u05D9\u05DE\u05D9\u05DD" },
    hour: { singular: "\u05E9\u05E2\u05D4", plural: "\u05E9\u05E2\u05D5\u05EA" },
    minute: { singular: "\u05D3\u05E7\u05D4", plural: "\u05D3\u05E7\u05D5\u05EA" },
    second: { singular: "\u05E9\u05E0\u05D9\u05D4", plural: "\u05E9\u05E0\u05D9\u05D5\u05EA" }
  },
  words: {
    every: "\u05DB\u05DC",
    weekday: "\u05D9\u05D5\u05DD \u05D7\u05D5\u05DC",
    on: "\u05D1",
    in: "\u05D1",
    on_the: "\u05D1",
    day_of_month: "\u05D9\u05D5\u05DD \u05D1\u05D7\u05D5\u05D3\u05E9",
    day_of_year: "\u05D9\u05D5\u05DD \u05D1\u05E9\u05E0\u05D4",
    in_week: "\u05D1\u05E9\u05D1\u05D5\u05E2",
    at: "\u05D1\u05E9\u05E2\u05D4",
    at_minute: "\u05D1\u05D3\u05E7\u05D4",
    at_second: "\u05D1\u05E9\u05E0\u05D9\u05D4",
    until: "\u05E2\u05D3",
    for: "\u05D1\u05DE\u05E9\u05DA",
    time: "\u05E4\u05E2\u05DD",
    times: "\u05E4\u05E2\u05DE\u05D9\u05DD",
    instance: "\u05E4\u05E2\u05DD",
    week_starts_on: "\u05E9\u05D1\u05D5\u05E2 \u05DE\u05EA\u05D7\u05D9\u05DC \u05D1",
    with: "\u05E2\u05DD",
    additional_date: "\u05EA\u05D0\u05E8\u05D9\u05DA \u05E0\u05D5\u05E1\u05E3",
    additional_dates: "\u05EA\u05D0\u05E8\u05D9\u05DB\u05D9\u05DD \u05E0\u05D5\u05E1\u05E4\u05D9\u05DD",
    excluding: "\u05DC\u05DE\u05E2\u05D8",
    date: "\u05EA\u05D0\u05E8\u05D9\u05DA",
    dates: "\u05EA\u05D0\u05E8\u05D9\u05DB\u05D9\u05DD",
    and: "\u05D5",
    last: "\u05D0\u05D7\u05E8\u05D5\u05DF",
    starting_from: "\u05D4\u05D7\u05DC \u05DE"
  },
  ordinal: (n) => {
    if (n < 0) return "\u05D0\u05D7\u05E8\u05D5\u05DF";
    const abs = Math.abs(n);
    const map = {
      1: "\u05E8\u05D0\u05E9\u05D5\u05DF",
      2: "\u05E9\u05E0\u05D9",
      3: "\u05E9\u05DC\u05D9\u05E9\u05D9",
      4: "\u05E8\u05D1\u05D9\u05E2\u05D9",
      5: "\u05D7\u05DE\u05D9\u05E9\u05D9",
      6: "\u05E9\u05D9\u05E9\u05D9",
      7: "\u05E9\u05D1\u05D9\u05E2\u05D9",
      8: "\u05E9\u05DE\u05D9\u05E0\u05D9",
      9: "\u05EA\u05E9\u05D9\u05E2\u05D9",
      10: "\u05E2\u05E9\u05D9\u05E8\u05D9",
      11: "\u05D0\u05D7\u05D3 \u05E2\u05E9\u05E8",
      12: "\u05E9\u05E0\u05D9\u05DD \u05E2\u05E9\u05E8",
      13: "\u05E9\u05DC\u05D5\u05E9\u05D4 \u05E2\u05E9\u05E8"
    };
    return map[abs] || abs.toString();
  }
};
var zh = {
  weekdayNames: ["\u661F\u671F\u4E00", "\u661F\u671F\u4E8C", "\u661F\u671F\u4E09", "\u661F\u671F\u56DB", "\u661F\u671F\u4E94", "\u661F\u671F\u516D", "\u661F\u671F\u65E5"],
  monthNames: ["\u4E00\u6708", "\u4E8C\u6708", "\u4E09\u6708", "\u56DB\u6708", "\u4E94\u6708", "\u516D\u6708", "\u4E03\u6708", "\u516B\u6708", "\u4E5D\u6708", "\u5341\u6708", "\u5341\u4E00\u6708", "\u5341\u4E8C\u6708"],
  units: {
    year: { singular: "\u5E74", plural: "\u5E74" },
    month: { singular: "\u6708", plural: "\u6708" },
    week: { singular: "\u5468", plural: "\u5468" },
    day: { singular: "\u65E5", plural: "\u65E5" },
    hour: { singular: "\u5C0F\u65F6", plural: "\u5C0F\u65F6" },
    minute: { singular: "\u5206\u949F", plural: "\u5206\u949F" },
    second: { singular: "\u79D2", plural: "\u79D2" }
  },
  words: {
    every: "\u6BCF",
    weekday: "\u5DE5\u4F5C\u65E5",
    on: "\u5728",
    in: "\u5728",
    on_the: "\u5728",
    day_of_month: "\u6708\u7684\u65E5\u5B50",
    day_of_year: "\u5E74\u7684\u65E5\u5B50",
    in_week: "\u7B2C",
    at: "\u5728",
    at_minute: "\u5728\u7B2C",
    at_second: "\u5728\u7B2C",
    until: "\u76F4\u5230",
    for: "\u5171",
    time: "\u6B21",
    times: "\u6B21",
    instance: "\u6B21",
    week_starts_on: "\u661F\u671F\u5F00\u59CB\u4E8E",
    with: "\u5E26\u6709",
    additional_date: "\u989D\u5916\u65E5\u671F",
    additional_dates: "\u989D\u5916\u65E5\u671F",
    excluding: "\u6392\u9664",
    date: "\u65E5\u671F",
    dates: "\u65E5\u671F",
    and: "\u548C",
    last: "\u6700\u540E",
    starting_from: "\u4ECE"
  },
  ordinal: (n) => n < 0 ? "\u6700\u540E" : `\u7B2C${Math.abs(n)}`
};
var fr = {
  weekdayNames: ["lundi", "mardi", "mercredi", "jeudi", "vendredi", "samedi", "dimanche"],
  monthNames: [
    "janvier",
    "f\xE9vrier",
    "mars",
    "avril",
    "mai",
    "juin",
    "juillet",
    "ao\xFBt",
    "septembre",
    "octobre",
    "novembre",
    "d\xE9cembre"
  ],
  units: {
    year: { singular: "ann\xE9e", plural: "ans" },
    month: { singular: "mois", plural: "mois" },
    week: { singular: "semaine", plural: "semaines" },
    day: { singular: "jour", plural: "jours" },
    hour: { singular: "heure", plural: "heures" },
    minute: { singular: "minute", plural: "minutes" },
    second: { singular: "seconde", plural: "secondes" }
  },
  words: {
    every: "chaque",
    weekday: "jour de semaine",
    on: "le",
    in: "dans",
    on_the: "le",
    day_of_month: "jour du mois",
    day_of_year: "jour de l'ann\xE9e",
    in_week: "dans la semaine",
    at: "\xE0",
    at_minute: "\xE0 la minute",
    at_second: "\xE0 la seconde",
    until: "jusqu'au",
    for: "pendant",
    time: "fois",
    times: "fois",
    instance: "occurrence",
    week_starts_on: "la semaine commence le",
    with: "avec",
    additional_date: "date suppl\xE9mentaire",
    additional_dates: "dates suppl\xE9mentaires",
    excluding: "en excluant",
    date: "date",
    dates: "dates",
    and: "et",
    last: "dernier",
    starting_from: "\xE0 partir du"
  },
  ordinal: (n) => {
    const abs = Math.abs(n);
    if (n < 0) return "dernier";
    if (abs === 1) return "1er";
    return `${abs}e`;
  }
};
var ALL_LOCALES = { en, de, es, hi, yue, ar, he, zh, fr };
var env = typeof process !== "undefined" && process.env ? process.env.TOTEXT_LANGS : void 0;
var active = env ? env.split(",").map((s) => s.trim()).filter(Boolean) : Object.keys(ALL_LOCALES);
var LOCALES = {};
for (const l of active) {
  if (ALL_LOCALES[l]) LOCALES[l] = ALL_LOCALES[l];
}
var workweekWeekdays = allowedWeekdays.slice(0, 5);
var byDayTokenRegex2 = new RegExp(`^([+-]?\\d+)?(${allowedWeekdays.join("|")})$`);
var weekdayIndexByToken = Object.fromEntries(
  allowedWeekdays.map((weekday, idx) => [weekday, idx])
);
function defaultOrdinal(n) {
  const abs = Math.abs(n);
  const suffix = abs % 10 === 1 && abs % 100 !== 11 ? "st" : abs % 10 === 2 && abs % 100 !== 12 ? "nd" : abs % 10 === 3 && abs % 100 !== 13 ? "rd" : "th";
  return n < 0 ? `last` : `${abs}${suffix}`;
}
function ordinal(n, locale) {
  return locale.ordinal ? locale.ordinal(n) : defaultOrdinal(n);
}
function list(arr, mapFn = (x) => `${x}`, final) {
  const mapped = arr.map(mapFn);
  if (mapped.length === 1) return mapped[0];
  return mapped.slice(0, -1).join(", ") + ` ${final} ` + mapped[mapped.length - 1];
}
function formatByDayToken(tok, locale) {
  if (typeof tok === "number") return tok.toString();
  const m = tok.toUpperCase().match(byDayTokenRegex2);
  if (!m) return tok;
  const ord = m[1] ? parseInt(m[1], 10) : 0;
  const weekday = m[2];
  const idx = weekdayIndexByToken[weekday];
  const name = locale.weekdayNames[idx];
  if (ord === 0) return name;
  if (ord === -1) return `${locale.words.last} ${name}`;
  return `${ordinal(ord, locale)} ${name}`;
}
function formatTime(hour, minute = 0, second = 0) {
  const hr12 = (hour + 11) % 12 + 1;
  const ampm = hour < 12 ? "AM" : "PM";
  const mm = String(minute).padStart(2, "0");
  const ss = String(second).padStart(2, "0");
  if (second) {
    return `${hr12}:${mm}:${ss} ${ampm}`;
  }
  if (minute) {
    return `${hr12}:${mm} ${ampm}`;
  }
  return `${hr12} ${ampm}`;
}
function weekdayTokenFromZdt(zdt) {
  return allowedWeekdays[zdt.dayOfWeek - 1];
}
function tzAbbreviation(zdt) {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: zdt.timeZoneId,
    timeZoneName: "short",
    hour: "numeric"
  }).formatToParts(new Date(zdt.epochMilliseconds));
  const tzPart = parts.find((p) => p.type === "timeZoneName");
  return (tzPart == null ? void 0 : tzPart.value) || zdt.timeZoneId;
}
function formatLocalizedDate(zdt, locale) {
  try {
    return zdt.toLocaleString(locale, { dateStyle: "long" });
  } catch (e) {
    return zdt.toLocaleString("en", { dateStyle: "long" });
  }
}
function toText(input, locale, options = {}) {
  const rule = typeof input === "string" ? new RRuleTemporal({ rruleString: input }) : input;
  const opts = rule.options();
  const resolvedLocale = locale != null ? locale : Intl.DateTimeFormat().resolvedOptions().locale;
  const lang = resolvedLocale.split("-")[0];
  const data = LOCALES[lang] || en;
  const dateLocale = LOCALES[lang] ? resolvedLocale : "en";
  const {
    freq,
    interval = 1,
    count,
    until,
    byDay,
    byHour,
    byMinute,
    bySecond,
    byMonth,
    byMonthDay,
    byYearDay,
    byWeekNo,
    bySetPos,
    wkst,
    rDate,
    exDate
  } = opts;
  const dayOrLarger = freq === "YEARLY" || freq === "MONTHLY" || freq === "WEEKLY" || freq === "DAILY";
  const hasExplicitTime = Boolean(byHour || byMinute || bySecond);
  const dtstartHasTime = opts.dtstart.hour !== 0 || opts.dtstart.minute !== 0 || opts.dtstart.second !== 0;
  const shouldDefaultTime = dayOrLarger && (hasExplicitTime || dtstartHasTime);
  const textByDay = byDay != null ? byDay : freq === "WEEKLY" ? [weekdayTokenFromZdt(opts.dtstart)] : void 0;
  const textByHour = byHour != null ? byHour : shouldDefaultTime ? [opts.dtstart.hour] : void 0;
  const textByMinute = textByHour ? byMinute != null ? byMinute : [opts.dtstart.minute] : byMinute;
  const textBySecond = textByHour ? bySecond != null ? bySecond : [opts.dtstart.second] : bySecond;
  const parts = [data.words.every];
  const baseKey = {
    YEARLY: "year",
    MONTHLY: "month",
    WEEKLY: "week",
    DAILY: "day",
    HOURLY: "hour",
    MINUTELY: "minute",
    SECONDLY: "second"
  }[freq];
  const base = data.units[baseKey];
  const daysNormalized = textByDay == null ? void 0 : textByDay.map((d) => d.toUpperCase());
  const isWeekdays = daysNormalized && daysNormalized.length === workweekWeekdays.length && workweekWeekdays.every((d) => daysNormalized.includes(d));
  const isEveryday = daysNormalized && daysNormalized.length === allowedWeekdays.length && allowedWeekdays.every((d) => daysNormalized.includes(d));
  if (freq === "WEEKLY" && interval === 1 && isWeekdays) {
    parts.push(data.words.weekday);
  } else if (freq === "WEEKLY" && interval === 1 && isEveryday) {
    parts.push(data.units.day.singular);
  } else {
    if (interval !== 1) {
      parts.push(interval.toString(), base.plural);
    } else {
      parts.push(base.singular);
    }
  }
  if (freq === "WEEKLY" && textByDay && !isWeekdays && !isEveryday) {
    parts.push(
      data.words.on,
      list(textByDay, (t) => formatByDayToken(t, data), data.words.and)
    );
  } else if (textByDay && freq !== "WEEKLY") {
    parts.push(
      data.words.on,
      list(textByDay, (t) => formatByDayToken(t, data), data.words.and)
    );
  }
  if (byMonth) {
    parts.push(
      data.words.in,
      list(byMonth, (m) => data.monthNames[m - 1], data.words.and)
    );
  }
  if (byMonthDay) {
    parts.push(
      data.words.on_the,
      list(byMonthDay, (d) => ordinal(d, data), data.words.and),
      data.words.day_of_month
    );
  }
  if (byYearDay) {
    parts.push(
      data.words.on_the,
      list(byYearDay, (d) => ordinal(d, data), data.words.and),
      data.words.day_of_year
    );
  }
  if (byWeekNo) {
    parts.push(
      data.words.in_week,
      list(byWeekNo, (n) => n.toString(), data.words.and)
    );
  }
  if (textByHour) {
    const minutes = textByMinute != null ? textByMinute : [0];
    const seconds = textBySecond != null ? textBySecond : [0];
    const times = textByHour.flatMap((h) => minutes.flatMap((m) => seconds.map((s) => formatTime(h, m, s))));
    parts.push(data.words.at, list(times, void 0, data.words.and));
    parts.push(tzAbbreviation(opts.dtstart));
  }
  if (!textByHour && textByMinute) {
    parts.push(data.words.at_minute, list(textByMinute, void 0, data.words.and));
  }
  if (!textByHour && !textByMinute && textBySecond) {
    parts.push(data.words.at_second, list(textBySecond, void 0, data.words.and));
  }
  if (options.includeDtstart) {
    parts.push(data.words.starting_from, formatLocalizedDate(opts.dtstart, dateLocale));
  }
  if (until) {
    parts.push(data.words.until, formatLocalizedDate(until, dateLocale));
  } else if (count !== void 0) {
    parts.push(data.words.for, count.toString(), count === 1 ? data.words.time : data.words.times);
  }
  if (bySetPos) {
    parts.push(
      data.words.on_the,
      list(bySetPos, (n) => ordinal(n, data), data.words.and),
      data.words.instance
    );
  }
  if (wkst) {
    const wkName = formatByDayToken(wkst, data);
    parts.push(data.words.week_starts_on, wkName);
  }
  if (rDate && rDate.length) {
    parts.push(
      data.words.with,
      `${rDate.length}`,
      rDate.length === 1 ? data.words.additional_date : data.words.additional_dates
    );
  }
  if (exDate && exDate.length) {
    parts.push(data.words.excluding, `${exDate.length}`, exDate.length === 1 ? data.words.date : data.words.dates);
  }
  if (opts.rscale) {
    const rscale = opts.rscale;
    const skip = opts.skip;
    parts.push(`(RSCALE=${rscale}${skip ? `;SKIP=${skip}` : ""})`);
  }
  return parts.join(" ");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  toText
});
